<!DOCTYPE html>

<html lang="en" data-content_root="./">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Simulation validation &#8212; Joint Modeling of SARS-CoV-2 Spike homologs 2023-07-21 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=d1102ebc" />
    <link rel="stylesheet" type="text/css" href="_static/alabaster.css?v=98334b14" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <script src="_static/documentation_options.js?v=04cdc583"></script>
    <script src="_static/doctools.js?v=888ff710"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  

  
  

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="simulation-validation">
<h1>Simulation validation<a class="headerlink" href="#simulation-validation" title="Link to this heading">¶</a></h1>
<section id="overview">
<h2>Overview<a class="headerlink" href="#overview" title="Link to this heading">¶</a></h2>
<p>The bloom lab had developed a <a class="reference external" href="https://jbloomlab.github.io/dms_variants/codonvariant_sim_data.html">pipeline to simulate DMS data</a> to test <code class="docutils literal notranslate"><span class="pre">dms_variants</span></code> global epistasis fitting. Taking inspriation from that, we build upon the pipeline by simulating data for <em>two</em> homologs - introducing shifts in mutational effects at a subset of sites with the goal validating the <code class="docutils literal notranslate"><span class="pre">multidms</span></code> joint-fitting approach.</p>
<p>This notebook has a few major steps involved:</p>
<ol class="arabic simple">
<li><p>Simulate sequences, and respective mutational effects for two homologs – introducing <em>shifts</em> at a subset of sites for the non-reference homolog.</p></li>
<li><p>Define replicate libraries of variants for each homolog.</p></li>
<li><p>Simulating pre, and post-selection library variant <em>counts</em> derived from a simulated bottleneck and used to introduce realistic noise in the fitting targets (functional scores).</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">multidms</span></code> model fitting.</p></li>
<li><p>Model fit comparison and selection criteria</p></li>
<li><p>Model evaluation.</p></li>
</ol>
</section>
<section id="import-python-modules">
<h2>Import <code class="docutils literal notranslate"><span class="pre">Python</span></code> modules<a class="headerlink" href="#import-python-modules" title="Link to this heading">¶</a></h2>
<p>For this analysis, we’ll be using the <code class="docutils literal notranslate"><span class="pre">dms_variants</span></code> package for simulation. All dependencies needed can be installed by cloning the <code class="docutils literal notranslate"><span class="pre">multidms</span></code> repo and executing the command:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ pip install &#39;.[dev]&#39;
</pre></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">warnings</span>
<span class="kn">import</span> <span class="nn">pprint</span>
<span class="kn">import</span> <span class="nn">os</span>

<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>

<span class="kn">import</span> <span class="nn">Bio.Seq</span>
<span class="kn">import</span> <span class="nn">itertools</span>
<span class="kn">import</span> <span class="nn">random</span>
<span class="kn">import</span> <span class="nn">dms_variants.codonvarianttable</span>
<span class="kn">import</span> <span class="nn">dms_variants.globalepistasis</span>
<span class="kn">import</span> <span class="nn">dms_variants.plotnine_themes</span>
<span class="kn">import</span> <span class="nn">dms_variants.simulate</span>
<span class="kn">from</span> <span class="nn">dms_variants.constants</span> <span class="kn">import</span> <span class="n">CBPALETTE</span><span class="p">,</span> <span class="n">CODONS_NOSTOP</span><span class="p">,</span> <span class="n">AAS_NOSTOP</span><span class="p">,</span> <span class="n">AA_TO_CODONS</span><span class="p">,</span> <span class="n">AAS_WITHSTOP</span>
<span class="kn">from</span> <span class="nn">plotnine</span> <span class="kn">import</span> <span class="o">*</span>
<span class="kn">import</span> <span class="nn">multidms</span>
<span class="kn">from</span> <span class="nn">collections</span> <span class="kn">import</span> <span class="n">defaultdict</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="define-notebook-parameters">
<h2>Define notebook parameters<a class="headerlink" href="#define-notebook-parameters" title="Link to this heading">¶</a></h2>
<p>Set parameters that define this notebook’s behavior. These can all be modifed (using <code class="docutils literal notranslate"><span class="pre">papermill</span></code>) to reasonable values and the notebook should execute as expected.</p>
<div class="cell tag_parameters docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">seed</span> <span class="o">=</span> <span class="mi">24</span>  <span class="c1"># random number seed</span>

<span class="c1"># define the reference protein sequence we&#39;re going to simulate</span>
<span class="n">genelength</span> <span class="o">=</span> <span class="mi">50</span>  <span class="c1"># gene length in codons</span>
<span class="n">wt_latent</span> <span class="o">=</span> <span class="mf">5.0</span> <span class="c1"># wildtype latent phenotype</span>
<span class="n">stop_effect</span> <span class="o">=</span> <span class="o">-</span><span class="mi">15</span> <span class="c1"># -15 is the default for `SigmoidPhenotypeSimulator` object</span>
<span class="n">norm_weights</span><span class="o">=</span><span class="p">((</span><span class="mf">0.4</span><span class="p">,</span> <span class="o">-</span><span class="mf">0.5</span><span class="p">,</span> <span class="mf">1.2</span><span class="p">),</span> <span class="p">(</span><span class="mf">0.6</span><span class="p">,</span> <span class="o">-</span><span class="mi">4</span><span class="p">,</span> <span class="mi">3</span><span class="p">))</span> <span class="c1"># See `SigmoidPhenotypeSimulator` for more details</span>

<span class="c1"># define the non reference sequence attributes.</span>
<span class="n">n_non_identical_sites</span> <span class="o">=</span> <span class="mi">10</span> <span class="c1"># number of amino-acid mutations separating homologs</span>
<span class="n">min_muteffect_in_bundle</span> <span class="o">=</span> <span class="o">-</span><span class="mf">1.0</span> <span class="c1"># minimum effect per mutation</span>
<span class="n">max_muteffect_in_bundle</span> <span class="o">=</span> <span class="mf">1.0</span> <span class="c1"># maximum effect per mutation</span>
<span class="n">n_shifted_non_identical_sites</span> <span class="o">=</span> <span class="mi">6</span> <span class="c1"># number of non identical sites (in the bundle) for which mutations at that site are expected to have shifted effects</span>
<span class="n">n_shifted_identical_sites</span> <span class="o">=</span> <span class="mi">4</span> <span class="c1"># number of amino-acid mutations separating shifted</span>
<span class="n">shift_gauss_variance</span> <span class="o">=</span> <span class="mf">1.0</span> <span class="c1"># variance of the gaussian distribution from which the shifted effects are drawn</span>

<span class="c1"># define the sigmoid phenotype parameters. See `multidms.biophysical.sigmoidal_global_epistasis` for more details.</span>
<span class="n">sigmoid_phenotype_scale</span> <span class="o">=</span> <span class="mi">6</span>

<span class="c1"># define the libraries</span>
<span class="n">variants_per_lib_genelength_scalar</span> <span class="o">=</span> <span class="mi">500</span>  <span class="c1"># variants per library</span>
<span class="n">avgmuts</span> <span class="o">=</span> <span class="mf">2.0</span>  <span class="c1"># average codon mutations per variant</span>
<span class="n">bclen</span> <span class="o">=</span> <span class="mi">16</span>  <span class="c1"># length of nucleotide barcode for each variant</span>
<span class="n">variant_error_rate</span> <span class="o">=</span> <span class="mf">0.0</span> <span class="c1">#0.005  # rate at which variant sequence mis-called</span>
<span class="n">avgdepth_per_variant</span> <span class="o">=</span> <span class="mi">200</span>  <span class="c1"># average per-variant sequencing depth</span>
<span class="n">lib_uniformity</span> <span class="o">=</span> <span class="mi">5</span>  <span class="c1"># uniformity of library pre-selection</span>
<span class="n">noise</span> <span class="o">=</span> <span class="mf">0.0</span> <span class="c1"># 0.05 # random gaussian noise added to the post-selection counts of the phenotype</span>

<span class="c1"># bottlenecks from pre- to post-selection</span>
<span class="n">bottleneck_variants_per_lib_scalar</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;tight_bottle&quot;</span><span class="p">:</span> <span class="mi">5</span><span class="p">,</span>
    <span class="s2">&quot;loose_bottle&quot;</span><span class="p">:</span> <span class="mi">100</span><span class="p">,</span>
<span class="p">}</span>

<span class="c1"># define the fitting parameters</span>
<span class="n">num_training_steps</span> <span class="o">=</span> <span class="mi">1</span>
<span class="n">iterations_per_step</span> <span class="o">=</span> <span class="mi">100000</span> <span class="c1"># default 20000</span>
<span class="n">scale_coeff_lasso_shift</span> <span class="o">=</span> <span class="p">[</span><span class="mf">0.0</span><span class="p">,</span> <span class="mf">5.00e-6</span><span class="p">,</span> <span class="mf">1.00e-05</span><span class="p">,</span> <span class="mf">2.00e-05</span><span class="p">,</span> <span class="mf">4.00e-05</span><span class="p">,</span> <span class="mf">8.00e-05</span><span class="p">,</span> <span class="mf">1.60e-04</span><span class="p">,</span> <span class="mf">3.20e-04</span><span class="p">,</span> <span class="mf">6.40e-04</span><span class="p">]</span> <span class="c1"># the sweep of lasso coefficient params</span>
<span class="n">init_beta_naught</span> <span class="o">=</span> <span class="mf">5.0</span> <span class="c1"># We&#39;ve found that we need to start with a higher beta_naught to get the model to converge correctly,</span>
<span class="n">scale_coeff_ridge_beta</span> <span class="o">=</span> <span class="mf">1e-7</span> <span class="c1"># the sweep of ridge coefficient params</span>
<span class="n">train_frac</span> <span class="o">=</span> <span class="mf">0.8</span> <span class="c1"># fraction of data to use for cross validation training.</span>
<span class="n">lasso_choice</span> <span class="o">=</span> <span class="mf">8e-5</span> <span class="c1"># the lasso coefficient to use for the final model</span>
<span class="n">csv_output_dir</span> <span class="o">=</span> <span class="s1">&#39;output&#39;</span> <span class="c1"># the directory to save the output csv files</span>
</pre></div>
</div>
</div>
</div>
<p>Set some configurations and a few other more experimental parameters which could break things if not set correctly:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%</span><span class="k">matplotlib</span> inline

<span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="n">seed</span><span class="p">)</span>
<span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="n">seed</span><span class="p">)</span>
<span class="n">warnings</span><span class="o">.</span><span class="n">simplefilter</span><span class="p">(</span><span class="s1">&#39;ignore&#39;</span><span class="p">)</span>

<span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">csv_output_dir</span><span class="p">):</span>
    <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">csv_output_dir</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="simulate-a-reference-sequence">
<h2>Simulate a reference sequence<a class="headerlink" href="#simulate-a-reference-sequence" title="Link to this heading">¶</a></h2>
<p>Start by simulating a sequence of nucleotides, then translate it to amino acids. We’ll use this sequence as the reference sequence for the simulated datasets.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">geneseq_h1</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">random</span><span class="o">.</span><span class="n">choices</span><span class="p">(</span><span class="n">CODONS_NOSTOP</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="n">genelength</span><span class="p">))</span>
<span class="n">aaseq_h1</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">Bio</span><span class="o">.</span><span class="n">Seq</span><span class="o">.</span><span class="n">Seq</span><span class="p">(</span><span class="n">geneseq_h1</span><span class="p">)</span><span class="o">.</span><span class="n">translate</span><span class="p">())</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Wildtype gene of </span><span class="si">{</span><span class="n">genelength</span><span class="si">}</span><span class="s2"> codons:</span><span class="se">\n</span><span class="si">{</span><span class="n">geneseq_h1</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Wildtype protein:</span><span class="se">\n</span><span class="si">{</span><span class="n">aaseq_h1</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Wildtype gene of 50 codons:
GGTTCCAGTTTTAGTGGAACCGTGAGCGGTGTAGTGCGGTCGTTTACTCTTTTCGTCCGTCAGTTCACCGTGTCCAGCCATCAGGACCCGTGCTGTCATTGCACAACCGGATACTATTTTTCGCAAGTCTGTACCCTGAGCAGACTCGGA
Wildtype protein:
GSSFSGTVSGVVRSFTLFVRQFTVSSHQDPCCHCTTGYYFSQVCTLSRLG
</pre></div>
</div>
</div>
</div>
</section>
<section id="mutational-effects">
<h2>Mutational effects<a class="headerlink" href="#mutational-effects" title="Link to this heading">¶</a></h2>
<p>Simulate latent mutational effects, $\beta_m$, storing data in a <code class="docutils literal notranslate"><span class="pre">SigmoidPhenotypeSimulator</span></code> object. Also create an identical object for the second homolog. Later, we’ll update this object to include shifted mutational effects.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mut_pheno_args</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;geneseq&quot;</span> <span class="p">:</span> <span class="n">geneseq_h1</span><span class="p">,</span>
    <span class="s2">&quot;wt_latent&quot;</span> <span class="p">:</span> <span class="n">wt_latent</span><span class="p">,</span>
    <span class="s2">&quot;seed&quot;</span> <span class="p">:</span> <span class="n">seed</span><span class="p">,</span>
    <span class="s2">&quot;stop_effect&quot;</span> <span class="p">:</span> <span class="n">stop_effect</span><span class="p">,</span>
    <span class="s2">&quot;norm_weights&quot;</span> <span class="p">:</span> <span class="n">norm_weights</span>
<span class="p">}</span>

<span class="n">SigmoidPhenotype_h1</span> <span class="o">=</span> <span class="n">dms_variants</span><span class="o">.</span><span class="n">simulate</span><span class="o">.</span><span class="n">SigmoidPhenotypeSimulator</span><span class="p">(</span><span class="o">**</span><span class="n">mut_pheno_args</span><span class="p">)</span>
<span class="n">SigmoidPhenotype_h2</span> <span class="o">=</span> <span class="n">dms_variants</span><span class="o">.</span><span class="n">simulate</span><span class="o">.</span><span class="n">SigmoidPhenotypeSimulator</span><span class="p">(</span><span class="o">**</span><span class="n">mut_pheno_args</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Plot the distribution of mutation effects for the first homolog.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">p</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">ggplot</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;muteffects&#39;</span><span class="p">:</span> <span class="n">SigmoidPhenotype_h1</span><span class="o">.</span><span class="n">muteffects</span><span class="p">}),</span>
           <span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="s1">&#39;muteffects&#39;</span><span class="p">))</span>
    <span class="o">+</span> <span class="n">geom_histogram</span><span class="p">(</span><span class="n">binwidth</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span> <span class="n">fill</span><span class="o">=</span><span class="s1">&#39;pink&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.7</span><span class="p">)</span>
    <span class="o">+</span> <span class="n">xlab</span><span class="p">(</span><span class="s2">&quot;mutation effect, $β$&quot;</span><span class="p">)</span>
    <span class="o">+</span> <span class="n">ylab</span><span class="p">(</span><span class="s1">&#39;number of mutations&#39;</span><span class="p">)</span>
    <span class="o">+</span> <span class="n">theme_classic</span><span class="p">()</span>
    <span class="o">+</span> <span class="n">theme</span><span class="p">(</span><span class="n">figure_size</span><span class="o">=</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">3</span><span class="p">))</span>
<span class="p">)</span>
<span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">(</span><span class="n">show</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/9e2c7f8196b5e4fef5105a098b6d695a1a7f1fe0b87be080eba0ac4605719c5c.png" src="_images/9e2c7f8196b5e4fef5105a098b6d695a1a7f1fe0b87be080eba0ac4605719c5c.png" />
</div>
</div>
<p>As expected, our mutational effects are weighted in the negative direction, as we expect the majority of mutations to be deleterious to a protein.</p>
</section>
<section id="non-reference-homolog-sequence">
<h2>Non-reference homolog sequence<a class="headerlink" href="#non-reference-homolog-sequence" title="Link to this heading">¶</a></h2>
<p>Next, we’ll simulate the DNA/protein sequence of second homolog by making a defined number of random amino-acid mutations to the first homolog. When choosing the “bundle” of mutation which separate the two homologs, we avoid mutations that decrease or increase the latent phenotype by more than <code class="docutils literal notranslate"><span class="pre">min_muteffect_in_bundle</span></code> and <code class="docutils literal notranslate"><span class="pre">max_muteffect_in_bundle</span></code> respectively. This is to ensure that the latent phenotype of the second homolog is not too different from the first homolog, an important assumption for the <code class="docutils literal notranslate"><span class="pre">multidms</span></code> model.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Input params</span>
<span class="n">non_identical_sites</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">random</span><span class="o">.</span><span class="n">sample</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">aaseq_h1</span><span class="p">)</span><span class="o">+</span><span class="mi">1</span><span class="p">),</span> <span class="n">n_non_identical_sites</span><span class="p">))</span>
<span class="n">aaseq_h2</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
<span class="n">geneseq_h2</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>


<span class="c1"># Iterate over each amino acid in the first homolog, and make a mutation if indicated.</span>
<span class="k">for</span> <span class="p">(</span><span class="n">aa_n</span><span class="p">,</span> <span class="n">aa</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">aaseq_h1</span><span class="p">,</span> <span class="mi">1</span><span class="p">):</span>
    <span class="n">codon</span> <span class="o">=</span> <span class="n">geneseq_h1</span><span class="p">[(</span><span class="n">aa_n</span><span class="o">-</span><span class="mi">1</span><span class="p">)</span><span class="o">*</span><span class="mi">3</span><span class="p">:</span><span class="mi">3</span><span class="o">*</span><span class="n">aa_n</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">aa_n</span> <span class="ow">in</span> <span class="n">non_identical_sites</span><span class="p">:</span>
        <span class="c1"># define the valid mutations for this site</span>
        <span class="n">valid_bundle_mutations</span> <span class="o">=</span> <span class="p">[</span>
            <span class="n">mut_aa</span> <span class="k">for</span> <span class="n">mut_aa</span> <span class="ow">in</span> <span class="n">AAS_NOSTOP</span>
            <span class="k">if</span> <span class="p">(</span>
                <span class="p">(</span><span class="n">mut_aa</span> <span class="o">!=</span> <span class="n">aa</span><span class="p">)</span> \
                <span class="ow">and</span> <span class="p">(</span><span class="n">SigmoidPhenotype_h1</span><span class="o">.</span><span class="n">muteffects</span><span class="p">[</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">aa</span><span class="si">}{</span><span class="n">aa_n</span><span class="si">}{</span><span class="n">mut_aa</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="n">min_muteffect_in_bundle</span><span class="p">)</span> \
                <span class="ow">and</span> <span class="p">(</span><span class="n">SigmoidPhenotype_h1</span><span class="o">.</span><span class="n">muteffects</span><span class="p">[</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">aa</span><span class="si">}{</span><span class="n">aa_n</span><span class="si">}{</span><span class="n">mut_aa</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="n">max_muteffect_in_bundle</span><span class="p">)</span>
            <span class="p">)</span>
        <span class="p">]</span>
        <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">valid_bundle_mutations</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">,</span> <span class="n">aa_n</span>
        <span class="n">mut_aa</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">valid_bundle_mutations</span><span class="p">)</span>
        <span class="n">aaseq_h2</span> <span class="o">+=</span> <span class="n">mut_aa</span>
        <span class="n">mut_codon</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">AA_TO_CODONS</span><span class="p">[</span><span class="n">mut_aa</span><span class="p">])</span>
        <span class="n">geneseq_h2</span> <span class="o">+=</span> <span class="n">mut_codon</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">aaseq_h2</span> <span class="o">+=</span> <span class="n">aa</span>
        <span class="n">geneseq_h2</span> <span class="o">+=</span> <span class="n">codon</span>

<span class="c1"># Store and summarize results</span>
<span class="n">homolog_seqs_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
    <span class="p">{</span><span class="s2">&quot;wt_aa_h1&quot;</span> <span class="p">:</span> <span class="nb">list</span><span class="p">(</span><span class="n">aaseq_h1</span><span class="p">),</span> <span class="s2">&quot;wt_aa_h2&quot;</span> <span class="p">:</span> <span class="nb">list</span><span class="p">(</span><span class="n">aaseq_h2</span><span class="p">)},</span>
    <span class="n">index</span> <span class="o">=</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">aaseq_h1</span><span class="p">)</span><span class="o">+</span><span class="mi">1</span><span class="p">),</span>
<span class="p">)</span>
<span class="n">homolog_seqs_df</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="s1">&#39;site&#39;</span>

<span class="c1"># n_diffs = sum([aa1 != aa2 for (aa1, aa2) in zip(*homolog_seqs.values())])</span>
<span class="n">n_diffs</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">homolog_seqs_df</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s1">&#39;wt_aa_h1 != wt_aa_h2&#39;</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Sequence alignment of homologs h1 and h2:&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;h1&#39;</span><span class="p">,</span> <span class="n">aaseq_h1</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;h2&#39;</span><span class="p">,</span> <span class="n">aaseq_h2</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Number of aa differences:&#39;</span><span class="p">,</span> <span class="n">n_diffs</span><span class="p">)</span>
<span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">aaseq_h1</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">aaseq_h2</span><span class="p">)</span>
<span class="k">assert</span> <span class="n">aaseq_h2</span> <span class="o">==</span> <span class="nb">str</span><span class="p">(</span><span class="n">Bio</span><span class="o">.</span><span class="n">Seq</span><span class="o">.</span><span class="n">Seq</span><span class="p">(</span><span class="n">geneseq_h2</span><span class="p">)</span><span class="o">.</span><span class="n">translate</span><span class="p">())</span>
<span class="k">assert</span> <span class="n">n_diffs</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">non_identical_sites</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Sequence alignment of homologs h1 and h2:
h1 GSSFSGTVSGVVRSFTLFVRQFTVSSHQDPCCHCTTGYYFSQVCTLSRLG
h2 GSSFSGTGEGVVRMFTLFVRQFTVSSHQDPCDHCITSDYFSHVETLSRVG
Number of aa differences: 10
</pre></div>
</div>
</div>
</div>
</section>
<section id="shifted-mutational-effects">
<h2>Shifted mutational effects<a class="headerlink" href="#shifted-mutational-effects" title="Link to this heading">¶</a></h2>
<p>Next, randomly choose a subset of (identical and non-identical) sites that will have shifted mutational effects. Do this independently for sites that are identical and non-identical between homologs, so that we are sure to have shifted sites in each category.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Non-identical sites</span>
<span class="n">shifted_non_identical_sites</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">random</span><span class="o">.</span><span class="n">sample</span><span class="p">(</span>
    <span class="n">non_identical_sites</span><span class="p">,</span>
    <span class="n">n_shifted_non_identical_sites</span>
<span class="p">))</span>

<span class="c1"># choose a subset of the identical sites to have shifted effects</span>
<span class="n">shifted_identical_sites</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">random</span><span class="o">.</span><span class="n">sample</span><span class="p">(</span>
    <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">aaseq_h1</span><span class="p">)</span><span class="o">+</span><span class="mi">1</span><span class="p">))</span> <span class="o">-</span> <span class="nb">set</span><span class="p">(</span><span class="n">non_identical_sites</span><span class="p">)),</span>
    <span class="n">n_shifted_identical_sites</span>
<span class="p">))</span>

<span class="c1"># Make a list of all shifted sites</span>
<span class="n">shifted_sites</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">shifted_identical_sites</span> <span class="o">+</span> <span class="n">shifted_non_identical_sites</span><span class="p">)</span>
<span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">shifted_sites</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">shifted_sites</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Sites with shifts that are...&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;identical (n=</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">shifted_identical_sites</span><span class="p">)</span><span class="si">}</span><span class="s1">):&#39;</span><span class="p">,</span> <span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="nb">str</span><span class="p">,</span> <span class="n">shifted_identical_sites</span><span class="p">)))</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;non-identical (n=</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">shifted_non_identical_sites</span><span class="p">)</span><span class="si">}</span><span class="s1">):&#39;</span><span class="p">,</span> <span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="nb">str</span><span class="p">,</span> <span class="n">shifted_non_identical_sites</span><span class="p">)))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Sites with shifts that are...
identical (n=4): 30, 34, 39, 48
non-identical (n=6): 8, 14, 32, 37, 38, 42
</pre></div>
</div>
</div>
</div>
<p>Next, we’ll create a dataframe for all mutational effects and shifts, simulating shifts at each of the above sites by randomly simulate a shift in the effect of each mutation by drawing shifts from a Gaussian distribution:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">sim_mut_shift</span><span class="p">(</span><span class="n">shifted_site</span><span class="p">,</span> <span class="n">mutation</span><span class="p">):</span>
    <span class="k">if</span> <span class="p">(</span><span class="ow">not</span> <span class="n">shifted_site</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="s1">&#39;*&#39;</span> <span class="ow">in</span> <span class="n">mutation</span><span class="p">):</span>
        <span class="k">return</span> <span class="mi">0</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">normal</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">scale</span><span class="o">=</span><span class="n">shift_gauss_variance</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>

<span class="c1"># Make a dataframe of mutation effects in the first homolog</span>
<span class="n">mut_effects_df</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="o">.</span><span class="n">from_dict</span><span class="p">(</span>
        <span class="n">SigmoidPhenotype_h1</span><span class="o">.</span><span class="n">muteffects</span><span class="p">,</span>
        <span class="n">orient</span><span class="o">=</span><span class="s1">&#39;index&#39;</span><span class="p">,</span>
        <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;beta_h1&#39;</span><span class="p">]</span>
    <span class="p">)</span>
    <span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
    <span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;index&#39;</span><span class="p">:</span><span class="s1">&#39;mutation&#39;</span><span class="p">})</span>
    <span class="o">.</span><span class="n">assign</span><span class="p">(</span>
        <span class="n">wt_aa</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;mutation&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">y</span><span class="p">:</span> <span class="n">multidms</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">split_sub</span><span class="p">(</span><span class="n">y</span><span class="p">)[</span><span class="mi">0</span><span class="p">]),</span>
        <span class="n">site</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;mutation&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">y</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="n">multidms</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">split_sub</span><span class="p">(</span><span class="n">y</span><span class="p">)[</span><span class="mi">1</span><span class="p">])),</span>
        <span class="n">mut_aa</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;mutation&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">y</span><span class="p">:</span> <span class="n">multidms</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">split_sub</span><span class="p">(</span><span class="n">y</span><span class="p">)[</span><span class="mi">2</span><span class="p">]),</span>
        <span class="n">shifted_site</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;site&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">shifted_sites</span><span class="p">),</span>
        <span class="n">shift</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
            <span class="k">lambda</span> <span class="n">row</span><span class="p">:</span> <span class="n">sim_mut_shift</span><span class="p">(</span>
                <span class="n">row</span><span class="p">[</span><span class="s1">&#39;shifted_site&#39;</span><span class="p">],</span>
                <span class="n">row</span><span class="p">[</span><span class="s1">&#39;mutation&#39;</span><span class="p">]</span>
            <span class="p">),</span>
            <span class="n">axis</span><span class="o">=</span><span class="mi">1</span>
        <span class="p">),</span>
        <span class="n">beta_h2</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;beta_h1&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;shift&#39;</span><span class="p">]</span>
    <span class="p">)</span>
    <span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">homolog_seqs_df</span><span class="p">,</span> <span class="n">left_on</span><span class="o">=</span><span class="s1">&#39;site&#39;</span><span class="p">,</span> <span class="n">right_index</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s1">&#39;left&#39;</span><span class="p">)</span>
    <span class="o">.</span><span class="n">assign</span><span class="p">(</span>
        <span class="n">bundle_mut</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;mut_aa&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;wt_aa_h2&#39;</span><span class="p">]</span>
    <span class="p">)</span>
<span class="p">)</span>

<span class="c1"># Show data for a subset of sites with shifts</span>
<span class="n">mut_effects_df</span><span class="p">[</span><span class="n">mut_effects_df</span><span class="p">[</span><span class="s1">&#39;shifted_site&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="kc">True</span><span class="p">][[</span>
    <span class="s1">&#39;site&#39;</span><span class="p">,</span> <span class="s1">&#39;wt_aa_h1&#39;</span><span class="p">,</span> <span class="s1">&#39;wt_aa_h2&#39;</span><span class="p">,</span> <span class="s1">&#39;mutation&#39;</span><span class="p">,</span>
    <span class="s1">&#39;beta_h1&#39;</span><span class="p">,</span> <span class="s1">&#39;shift&#39;</span><span class="p">,</span> <span class="s1">&#39;beta_h2&#39;</span><span class="p">,</span>
    <span class="s1">&#39;shifted_site&#39;</span>
<span class="p">]]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>site</th>
      <th>wt_aa_h1</th>
      <th>wt_aa_h2</th>
      <th>mutation</th>
      <th>beta_h1</th>
      <th>shift</th>
      <th>beta_h2</th>
      <th>shifted_site</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>140</th>
      <td>8</td>
      <td>V</td>
      <td>G</td>
      <td>V8A</td>
      <td>-7.389801</td>
      <td>1.329212</td>
      <td>-6.060588</td>
      <td>True</td>
    </tr>
    <tr>
      <th>141</th>
      <td>8</td>
      <td>V</td>
      <td>G</td>
      <td>V8C</td>
      <td>-7.620390</td>
      <td>-0.770033</td>
      <td>-8.390423</td>
      <td>True</td>
    </tr>
    <tr>
      <th>142</th>
      <td>8</td>
      <td>V</td>
      <td>G</td>
      <td>V8D</td>
      <td>-7.295378</td>
      <td>-0.316280</td>
      <td>-7.611658</td>
      <td>True</td>
    </tr>
    <tr>
      <th>143</th>
      <td>8</td>
      <td>V</td>
      <td>G</td>
      <td>V8E</td>
      <td>-0.729143</td>
      <td>-0.990810</td>
      <td>-1.719953</td>
      <td>True</td>
    </tr>
    <tr>
      <th>144</th>
      <td>8</td>
      <td>V</td>
      <td>G</td>
      <td>V8F</td>
      <td>-1.369987</td>
      <td>-1.070816</td>
      <td>-2.440803</td>
      <td>True</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>955</th>
      <td>48</td>
      <td>R</td>
      <td>R</td>
      <td>R48T</td>
      <td>-6.008805</td>
      <td>-2.870096</td>
      <td>-8.878901</td>
      <td>True</td>
    </tr>
    <tr>
      <th>956</th>
      <td>48</td>
      <td>R</td>
      <td>R</td>
      <td>R48V</td>
      <td>0.899548</td>
      <td>0.351710</td>
      <td>1.251258</td>
      <td>True</td>
    </tr>
    <tr>
      <th>957</th>
      <td>48</td>
      <td>R</td>
      <td>R</td>
      <td>R48W</td>
      <td>-8.789600</td>
      <td>0.265490</td>
      <td>-8.524110</td>
      <td>True</td>
    </tr>
    <tr>
      <th>958</th>
      <td>48</td>
      <td>R</td>
      <td>R</td>
      <td>R48Y</td>
      <td>-5.326602</td>
      <td>0.105712</td>
      <td>-5.220890</td>
      <td>True</td>
    </tr>
    <tr>
      <th>959</th>
      <td>48</td>
      <td>R</td>
      <td>R</td>
      <td>R48*</td>
      <td>-15.000000</td>
      <td>0.000000</td>
      <td>-15.000000</td>
      <td>True</td>
    </tr>
  </tbody>
</table>
<p>200 rows × 8 columns</p>
</div></div></div>
</div>
<p>Plot the distribution of all simulated shifts, excluding mutations to stop codons and sites that have no shifted effects.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">non_stop_shifted_muts</span> <span class="o">=</span> <span class="n">mut_effects_df</span><span class="p">[</span>
    <span class="p">(</span><span class="n">mut_effects_df</span><span class="p">[</span><span class="s1">&#39;site&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">shifted_sites</span><span class="p">))</span> <span class="o">&amp;</span>
    <span class="o">~</span><span class="p">(</span><span class="n">mut_effects_df</span><span class="p">[</span><span class="s1">&#39;mutation&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s1">&#39;\*&#39;</span><span class="p">))</span>
<span class="p">]</span>

<span class="n">p</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">ggplot</span><span class="p">(</span><span class="n">non_stop_shifted_muts</span><span class="p">,</span> <span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="s1">&#39;shift&#39;</span><span class="p">))</span>
    <span class="o">+</span> <span class="n">geom_histogram</span><span class="p">(</span><span class="n">binwidth</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span> <span class="n">fill</span><span class="o">=</span><span class="s1">&#39;pink&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.7</span><span class="p">)</span>
    <span class="o">+</span> <span class="n">xlab</span><span class="p">(</span><span class="s2">&quot;shift&quot;</span><span class="p">)</span>
    <span class="o">+</span> <span class="n">ylab</span><span class="p">(</span><span class="s1">&#39;number of mutations&#39;</span><span class="p">)</span>
    <span class="o">+</span> <span class="n">theme_classic</span><span class="p">()</span>
    <span class="o">+</span> <span class="n">theme</span><span class="p">(</span><span class="n">figure_size</span><span class="o">=</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">3</span><span class="p">))</span>
<span class="p">)</span>

<span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">(</span><span class="n">show</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/fd34b57659d3bdaaaca51274626b6ed8d80b5257def6ced955f15f3dc3c44989.png" src="_images/fd34b57659d3bdaaaca51274626b6ed8d80b5257def6ced955f15f3dc3c44989.png" />
</div>
</div>
<p>We can also plot the shifts as a heatmap, and mark the wildtype amino acid at each site (<code class="docutils literal notranslate"><span class="pre">x</span></code> for reference, <code class="docutils literal notranslate"><span class="pre">o</span></code> for non-reference):</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">max_abs_value</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="nb">abs</span><span class="p">(</span><span class="n">mut_effects_df</span><span class="p">[</span><span class="s1">&#39;shift&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">min</span><span class="p">()),</span> <span class="n">mut_effects_df</span><span class="p">[</span><span class="s1">&#39;shift&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">max</span><span class="p">())</span>
<span class="n">x_scale</span> <span class="o">=</span> <span class="p">[</span><span class="n">s</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">homolog_seqs_df</span><span class="p">)</span><span class="o">+</span><span class="mi">1</span><span class="p">)]</span>
<span class="n">y_scale</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">([</span>
    <span class="s2">&quot;R&quot;</span><span class="p">,</span><span class="s2">&quot;K&quot;</span><span class="p">,</span><span class="s2">&quot;H&quot;</span><span class="p">,</span><span class="s2">&quot;D&quot;</span><span class="p">,</span><span class="s2">&quot;E&quot;</span><span class="p">,</span><span class="s2">&quot;Q&quot;</span><span class="p">,</span><span class="s2">&quot;N&quot;</span><span class="p">,</span><span class="s2">&quot;S&quot;</span><span class="p">,</span>
    <span class="s2">&quot;T&quot;</span><span class="p">,</span><span class="s2">&quot;Y&quot;</span><span class="p">,</span><span class="s2">&quot;W&quot;</span><span class="p">,</span><span class="s2">&quot;F&quot;</span><span class="p">,</span><span class="s2">&quot;A&quot;</span><span class="p">,</span><span class="s2">&quot;I&quot;</span><span class="p">,</span><span class="s2">&quot;L&quot;</span><span class="p">,</span><span class="s2">&quot;M&quot;</span><span class="p">,</span>
    <span class="s2">&quot;V&quot;</span><span class="p">,</span><span class="s2">&quot;G&quot;</span><span class="p">,</span><span class="s2">&quot;P&quot;</span><span class="p">,</span><span class="s2">&quot;C&quot;</span><span class="p">,</span><span class="s2">&quot;*&quot;</span><span class="p">,</span>
<span class="p">])</span>

<span class="n">p</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">ggplot</span><span class="p">(</span><span class="n">mut_effects_df</span><span class="p">)</span>
    <span class="o">+</span> <span class="n">geom_tile</span><span class="p">(</span>
        <span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="s1">&#39;factor(site)&#39;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s1">&#39;mut_aa&#39;</span><span class="p">,</span> <span class="n">fill</span><span class="o">=</span><span class="s1">&#39;shift&#39;</span><span class="p">,</span> <span class="n">width</span><span class="o">=</span><span class="mf">.9</span><span class="p">,</span> <span class="n">height</span><span class="o">=</span><span class="mf">.9</span><span class="p">)</span>
    <span class="p">)</span>
    <span class="o">+</span> <span class="n">geom_point</span><span class="p">(</span>
        <span class="n">homolog_seqs_df</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(),</span>
        <span class="n">aes</span><span class="p">(</span>
            <span class="n">x</span><span class="o">=</span><span class="s1">&#39;factor(site)&#39;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s1">&#39;wt_aa_h1&#39;</span> <span class="c1">#, label=&#39;aaseq_h1&#39;</span>
        <span class="p">),</span>
        <span class="n">color</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span>
        <span class="n">size</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
        <span class="n">shape</span><span class="o">=</span><span class="s1">&#39;x&#39;</span>
    <span class="p">)</span>
    <span class="o">+</span> <span class="n">geom_point</span><span class="p">(</span>
        <span class="n">homolog_seqs_df</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s1">&#39;wt_aa_h1 != wt_aa_h2&#39;</span><span class="p">),</span>
        <span class="n">aes</span><span class="p">(</span>
            <span class="n">x</span><span class="o">=</span><span class="s1">&#39;factor(site)&#39;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s1">&#39;wt_aa_h2&#39;</span>
        <span class="p">),</span>
        <span class="n">color</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span>
        <span class="n">size</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
        <span class="n">shape</span><span class="o">=</span><span class="s1">&#39;o&#39;</span>
    <span class="p">)</span>
    <span class="o">+</span> <span class="n">scale_x_discrete</span><span class="p">(</span>
        <span class="n">limits</span> <span class="o">=</span> <span class="n">x_scale</span><span class="p">,</span>
        <span class="n">labels</span> <span class="o">=</span> <span class="n">x_scale</span>
    <span class="p">)</span>
    <span class="o">+</span> <span class="n">scale_y_discrete</span><span class="p">(</span>
        <span class="n">limits</span> <span class="o">=</span> <span class="n">y_scale</span><span class="p">,</span>
        <span class="n">labels</span> <span class="o">=</span> <span class="n">y_scale</span>
    <span class="p">)</span>
    <span class="o">+</span> <span class="n">scale_fill_gradientn</span><span class="p">(</span>
        <span class="n">colors</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;red&#39;</span><span class="p">,</span> <span class="s1">&#39;white&#39;</span><span class="p">,</span> <span class="s1">&#39;blue&#39;</span><span class="p">],</span> 
        <span class="n">limits</span><span class="o">=</span><span class="p">[</span><span class="o">-</span><span class="n">max_abs_value</span><span class="p">,</span> <span class="n">max_abs_value</span><span class="p">],</span>
    <span class="p">)</span>
    <span class="o">+</span> <span class="n">theme_classic</span><span class="p">()</span>
    <span class="o">+</span> <span class="n">theme</span><span class="p">(</span>
        <span class="n">axis_text_x</span><span class="o">=</span><span class="n">element_text</span><span class="p">(</span><span class="n">angle</span><span class="o">=</span><span class="mi">90</span><span class="p">),</span>
        <span class="n">figure_size</span><span class="o">=</span><span class="p">(</span><span class="mi">7</span><span class="p">,</span> <span class="mf">3.5</span><span class="p">)</span>
    <span class="p">)</span>
    <span class="o">+</span> <span class="n">xlab</span><span class="p">(</span><span class="s1">&#39;site&#39;</span><span class="p">)</span>
    <span class="o">+</span> <span class="n">ylab</span><span class="p">(</span><span class="s1">&#39;mutation&#39;</span><span class="p">)</span>
<span class="p">)</span>
<span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">(</span><span class="n">show</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/9c82e73c91367a626085b0b68485470c54630e4f8e2ee2cfadfe233d9a821da7.png" src="_images/9c82e73c91367a626085b0b68485470c54630e4f8e2ee2cfadfe233d9a821da7.png" />
</div>
</div>
<p>Recal that we have already created a <code class="docutils literal notranslate"><span class="pre">SigmoidPhenotypeSimulator</span></code> object for the second homolog by making a copy of the one for the first homolog. This object stores the homolog’s wildtype latent phenotype and the latent effects of individual mutations. Below, we update both of these traits based on the simulated shifts from above. To update the wildtype latent phenotype of the second homolog, we add the effects of all mutations that separate the homologs. In computing this sum, we will use $\beta$ parameters for the second homolog, which already include shifted effects.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Update individual mutational effects</span>
<span class="k">assert</span> <span class="nb">sum</span><span class="p">(</span><span class="n">mut_effects_df</span><span class="p">[</span><span class="s1">&#39;mutation&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">duplicated</span><span class="p">())</span> <span class="o">==</span> <span class="mi">0</span>
<span class="k">for</span> <span class="n">mutation</span> <span class="ow">in</span> <span class="n">SigmoidPhenotype_h2</span><span class="o">.</span><span class="n">muteffects</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
    <span class="n">SigmoidPhenotype_h2</span><span class="o">.</span><span class="n">muteffects</span><span class="p">[</span><span class="n">mutation</span><span class="p">]</span> <span class="o">=</span> <span class="n">mut_effects_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span>
        <span class="n">mut_effects_df</span><span class="p">[</span><span class="s1">&#39;mutation&#39;</span><span class="p">]</span><span class="o">==</span><span class="n">mutation</span><span class="p">,</span> <span class="s1">&#39;beta_h2&#39;</span>
    <span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>

<span class="n">wt_latent_phenotype_shift</span> <span class="o">=</span> <span class="n">mut_effects_df</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s2">&quot;bundle_mut&quot;</span><span class="p">)[</span><span class="s2">&quot;beta_h2&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
<span class="n">SigmoidPhenotype_h2</span><span class="o">.</span><span class="n">wt_latent</span> <span class="o">=</span> <span class="n">SigmoidPhenotype_h1</span><span class="o">.</span><span class="n">wt_latent</span> <span class="o">+</span> <span class="n">wt_latent_phenotype_shift</span>

<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Characteristics of mutations separating homologs:</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">metric</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;beta_h1&#39;</span><span class="p">,</span> <span class="s1">&#39;shift&#39;</span><span class="p">,</span> <span class="s1">&#39;beta_h2&#39;</span><span class="p">]:</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;  Sum of </span><span class="si">{</span><span class="n">metric</span><span class="si">}</span><span class="s1">:&#39;</span><span class="p">,</span> <span class="nb">round</span><span class="p">(</span><span class="nb">sum</span><span class="p">(</span><span class="n">mut_effects_df</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s2">&quot;bundle_mut&quot;</span><span class="p">)[</span><span class="n">metric</span><span class="p">]),</span> <span class="mi">2</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;  Final WT latent phenotype of h2:&#39;</span><span class="p">,</span> <span class="nb">round</span><span class="p">(</span><span class="n">SigmoidPhenotype_h2</span><span class="o">.</span><span class="n">wt_latent</span><span class="p">,</span> <span class="mi">2</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;  Table of mutations that separate homologs:&#39;</span><span class="p">)</span>
<span class="n">mut_effects_df</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s2">&quot;mut_aa == wt_aa_h2&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Characteristics of mutations separating homologs:

  Sum of beta_h1: 0.66
  Sum of shift: -0.28
  Sum of beta_h2: 0.38
  Final WT latent phenotype of h2: 5.38
  Table of mutations that separate homologs:
</pre></div>
</div>
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>mutation</th>
      <th>beta_h1</th>
      <th>wt_aa</th>
      <th>site</th>
      <th>mut_aa</th>
      <th>shifted_site</th>
      <th>shift</th>
      <th>beta_h2</th>
      <th>wt_aa_h1</th>
      <th>wt_aa_h2</th>
      <th>bundle_mut</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>145</th>
      <td>V8G</td>
      <td>0.834349</td>
      <td>V</td>
      <td>8</td>
      <td>G</td>
      <td>True</td>
      <td>-1.438713</td>
      <td>-0.604364</td>
      <td>V</td>
      <td>G</td>
      <td>True</td>
    </tr>
    <tr>
      <th>163</th>
      <td>S9E</td>
      <td>-0.705146</td>
      <td>S</td>
      <td>9</td>
      <td>E</td>
      <td>False</td>
      <td>0.000000</td>
      <td>-0.705146</td>
      <td>S</td>
      <td>E</td>
      <td>True</td>
    </tr>
    <tr>
      <th>270</th>
      <td>S14M</td>
      <td>0.192719</td>
      <td>S</td>
      <td>14</td>
      <td>M</td>
      <td>True</td>
      <td>1.037528</td>
      <td>1.230247</td>
      <td>S</td>
      <td>M</td>
      <td>True</td>
    </tr>
    <tr>
      <th>621</th>
      <td>C32D</td>
      <td>0.958833</td>
      <td>C</td>
      <td>32</td>
      <td>D</td>
      <td>True</td>
      <td>-0.538963</td>
      <td>0.419870</td>
      <td>C</td>
      <td>D</td>
      <td>True</td>
    </tr>
    <tr>
      <th>687</th>
      <td>T35I</td>
      <td>0.097535</td>
      <td>T</td>
      <td>35</td>
      <td>I</td>
      <td>False</td>
      <td>0.000000</td>
      <td>0.097535</td>
      <td>T</td>
      <td>I</td>
      <td>True</td>
    </tr>
    <tr>
      <th>734</th>
      <td>G37S</td>
      <td>-0.695803</td>
      <td>G</td>
      <td>37</td>
      <td>S</td>
      <td>True</td>
      <td>0.990621</td>
      <td>0.294818</td>
      <td>G</td>
      <td>S</td>
      <td>True</td>
    </tr>
    <tr>
      <th>742</th>
      <td>Y38D</td>
      <td>-0.265583</td>
      <td>Y</td>
      <td>38</td>
      <td>D</td>
      <td>True</td>
      <td>0.669489</td>
      <td>0.403906</td>
      <td>Y</td>
      <td>D</td>
      <td>True</td>
    </tr>
    <tr>
      <th>826</th>
      <td>Q42H</td>
      <td>-0.888294</td>
      <td>Q</td>
      <td>42</td>
      <td>H</td>
      <td>True</td>
      <td>-0.996761</td>
      <td>-1.885055</td>
      <td>Q</td>
      <td>H</td>
      <td>True</td>
    </tr>
    <tr>
      <th>862</th>
      <td>C44E</td>
      <td>0.136021</td>
      <td>C</td>
      <td>44</td>
      <td>E</td>
      <td>False</td>
      <td>0.000000</td>
      <td>0.136021</td>
      <td>C</td>
      <td>E</td>
      <td>True</td>
    </tr>
    <tr>
      <th>976</th>
      <td>L49V</td>
      <td>0.996403</td>
      <td>L</td>
      <td>49</td>
      <td>V</td>
      <td>False</td>
      <td>0.000000</td>
      <td>0.996403</td>
      <td>L</td>
      <td>V</td>
      <td>True</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>At this point, all mutations in the second homolog’s <code class="docutils literal notranslate"><span class="pre">SigmoidPhenotypeSimulator</span></code> object are defined relative to the wildtype sequence of the first homolog, which we will call the “reference” homolog. However, to compute phenotypes for the second homolog, we still need to assign the effects of mutations at non-identical sites in the “non-reference” homolog. For instance, if the wildtype amino acid at site 30 is an A in the reference homolog, but is a Y in a non-reference homolog, then the effect of a Y30G is absent from the second homolog’s simulator object.</p>
<p>To add these missing entries, we will use a strategy that assumes additivity between mutations at the same site.
For instance, in the above example, the effect of a Y30G mutation is defined as the sum of the A30Y (negated) and A30G effects in the reference homolog.
This approach assumes that mutational effects can be negated, such that A30Y has the opposite effect as Y30A.
It also assumes that mutational effects are additive, such that the effect of Y30G is the sum of the effects of Y30A and A30G.</p>
<p>The below expression defines this approach more explicitly for an arbitrary site.
It uses the notation $\beta_{x,n,z}$ where $x$ and $z$ are amino acids.
For a site $n$, if $aa_{\text{wt}}$ is the site’s wildtype amino acid in a non-reference homolog, $aa_{\text{mut}}$ is a mutant amino acid in a variant of that homolog, and $aa_{\text{ref}}$ is the site’s wildtype amino acid in the reference homolog, then the mutation’s effect is:</p>
<p>$$\beta_{aa_{\text{wt}},n,aa_{\text{mut}}} = \begin{cases}
\beta_{aa_{\text{ref}},n,aa_{\text{mut}}} &amp; \text{if } aa_{\text{wt}} = aa_{\text{ref}}\
-\beta_{aa_{\text{ref}},n,aa_{\text{wt}}} &amp; \text{if } aa_{\text{mut}} = aa_{\text{ref}}\
- \beta_{aa_{\text{ref}},n,aa_{\text{wt}}} +\beta_{aa_{\text{ref}},n,aa_{\text{mut}}} &amp; \text{otherwise}\
\end{cases}$$</p>
<p>The above $\beta$ parameters are effects in the background of the non-reference homolog, such that they account for shifts in mutational effects between homologs.</p>
<p>The below cell adds mutational effects for missing entries using the above strategy:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># iterate over &quot;bundle&quot; of mutations which disguish the homologs</span>
<span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">mut_effects_df</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s2">&quot;mut_aa == wt_aa_h2&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>

    <span class="c1"># compute the effect of each possible mutation at a bundle site</span>
    <span class="k">for</span> <span class="n">aa_mut</span> <span class="ow">in</span> <span class="n">AAS_WITHSTOP</span><span class="p">:</span>

        <span class="c1"># skip the non-ref wt-to-wt</span>
        <span class="k">if</span> <span class="n">aa_mut</span> <span class="o">==</span> <span class="n">row</span><span class="o">.</span><span class="n">wt_aa_h2</span><span class="p">:</span>
            <span class="k">continue</span>

        <span class="n">non_ref_mutation</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">row</span><span class="o">.</span><span class="n">wt_aa_h2</span><span class="si">}{</span><span class="n">row</span><span class="o">.</span><span class="n">site</span><span class="si">}{</span><span class="n">aa_mut</span><span class="si">}</span><span class="s1">&#39;</span>

        <span class="c1"># stop effect is the same for all homologs</span>
        <span class="k">if</span> <span class="n">aa_mut</span> <span class="o">==</span> <span class="s1">&#39;*&#39;</span><span class="p">:</span>
            <span class="n">SigmoidPhenotype_h2</span><span class="o">.</span><span class="n">muteffects</span><span class="p">[</span><span class="n">non_ref_mutation</span><span class="p">]</span> <span class="o">=</span> <span class="n">stop_effect</span>

        <span class="c1"># back-mutation to reference        </span>
        <span class="k">elif</span> <span class="n">aa_mut</span> <span class="o">==</span> <span class="n">row</span><span class="o">.</span><span class="n">wt_aa_h1</span><span class="p">:</span>
            <span class="n">SigmoidPhenotype_h2</span><span class="o">.</span><span class="n">muteffects</span><span class="p">[</span><span class="n">non_ref_mutation</span><span class="p">]</span> <span class="o">=</span> <span class="o">-</span><span class="n">row</span><span class="o">.</span><span class="n">beta_h2</span>

        <span class="c1"># all other mutations</span>
        <span class="k">else</span><span class="p">:</span>            
            <span class="n">ref_mut</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">row</span><span class="o">.</span><span class="n">wt_aa_h1</span><span class="si">}{</span><span class="n">row</span><span class="o">.</span><span class="n">site</span><span class="si">}{</span><span class="n">aa_mut</span><span class="si">}</span><span class="s1">&#39;</span>
            <span class="n">ref_mut_effect</span> <span class="o">=</span> <span class="n">mut_effects_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span>
                <span class="n">mut_effects_df</span><span class="p">[</span><span class="s1">&#39;mutation&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">ref_mut</span><span class="p">,</span> <span class="s1">&#39;beta_h2&#39;</span>
            <span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">SigmoidPhenotype_h2</span><span class="o">.</span><span class="n">muteffects</span><span class="p">[</span><span class="n">non_ref_mutation</span><span class="p">]</span> <span class="o">=</span> <span class="o">-</span><span class="n">row</span><span class="o">.</span><span class="n">beta_h2</span> <span class="o">+</span> <span class="n">ref_mut_effect</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mut_effects_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">csv_output_dir</span><span class="si">}</span><span class="s2">/simulated_muteffects.csv&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">mut_effects_df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>mutation</th>
      <th>beta_h1</th>
      <th>wt_aa</th>
      <th>site</th>
      <th>mut_aa</th>
      <th>shifted_site</th>
      <th>shift</th>
      <th>beta_h2</th>
      <th>wt_aa_h1</th>
      <th>wt_aa_h2</th>
      <th>bundle_mut</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>G1A</td>
      <td>-2.981264</td>
      <td>G</td>
      <td>1</td>
      <td>A</td>
      <td>False</td>
      <td>0.0</td>
      <td>-2.981264</td>
      <td>G</td>
      <td>G</td>
      <td>False</td>
    </tr>
    <tr>
      <th>1</th>
      <td>G1C</td>
      <td>-5.609745</td>
      <td>G</td>
      <td>1</td>
      <td>C</td>
      <td>False</td>
      <td>0.0</td>
      <td>-5.609745</td>
      <td>G</td>
      <td>G</td>
      <td>False</td>
    </tr>
    <tr>
      <th>2</th>
      <td>G1D</td>
      <td>-0.751102</td>
      <td>G</td>
      <td>1</td>
      <td>D</td>
      <td>False</td>
      <td>0.0</td>
      <td>-0.751102</td>
      <td>G</td>
      <td>G</td>
      <td>False</td>
    </tr>
    <tr>
      <th>3</th>
      <td>G1E</td>
      <td>-5.157105</td>
      <td>G</td>
      <td>1</td>
      <td>E</td>
      <td>False</td>
      <td>0.0</td>
      <td>-5.157105</td>
      <td>G</td>
      <td>G</td>
      <td>False</td>
    </tr>
    <tr>
      <th>4</th>
      <td>G1F</td>
      <td>-1.019324</td>
      <td>G</td>
      <td>1</td>
      <td>F</td>
      <td>False</td>
      <td>0.0</td>
      <td>-1.019324</td>
      <td>G</td>
      <td>G</td>
      <td>False</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
</section>
<section id="variant-libraries">
<h2>Variant libraries<a class="headerlink" href="#variant-libraries" title="Link to this heading">¶</a></h2>
<p>Simulate a set of variant libraries that one might use in an actual experiment for each homolog, each with two replicate libraries.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">libs</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;lib_1&quot;</span><span class="p">,</span> <span class="s2">&quot;lib_2&quot;</span><span class="p">]</span>  <span class="c1"># distinct libraries of gene</span>
<span class="n">variants_per_lib</span> <span class="o">=</span> <span class="n">variants_per_lib_genelength_scalar</span> <span class="o">*</span> <span class="n">genelength</span>

<span class="n">CodonVariantTable_h1</span> <span class="o">=</span> <span class="n">dms_variants</span><span class="o">.</span><span class="n">simulate</span><span class="o">.</span><span class="n">simulate_CodonVariantTable</span><span class="p">(</span>
    <span class="n">geneseq</span><span class="o">=</span><span class="n">geneseq_h1</span><span class="p">,</span>
    <span class="n">bclen</span><span class="o">=</span><span class="n">bclen</span><span class="p">,</span>
    <span class="n">library_specs</span><span class="o">=</span><span class="p">{</span>
        <span class="n">lib</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;avgmuts&quot;</span><span class="p">:</span> <span class="n">avgmuts</span><span class="p">,</span> <span class="s2">&quot;nvariants&quot;</span><span class="p">:</span> <span class="n">variants_per_lib</span><span class="p">}</span> <span class="k">for</span> <span class="n">lib</span> <span class="ow">in</span> <span class="n">libs</span>
    <span class="p">},</span>
    <span class="n">seed</span><span class="o">=</span><span class="n">seed</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">CodonVariantTable_h2</span> <span class="o">=</span> <span class="n">dms_variants</span><span class="o">.</span><span class="n">simulate</span><span class="o">.</span><span class="n">simulate_CodonVariantTable</span><span class="p">(</span>
    <span class="n">geneseq</span><span class="o">=</span><span class="n">geneseq_h2</span><span class="p">,</span>
    <span class="n">bclen</span><span class="o">=</span><span class="n">bclen</span><span class="p">,</span>
    <span class="n">library_specs</span><span class="o">=</span><span class="p">{</span>
        <span class="n">lib</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;avgmuts&quot;</span><span class="p">:</span> <span class="n">avgmuts</span><span class="p">,</span> <span class="s2">&quot;nvariants&quot;</span><span class="p">:</span> <span class="n">variants_per_lib</span><span class="p">}</span> <span class="k">for</span> <span class="n">lib</span> <span class="ow">in</span> <span class="n">libs</span>
    <span class="p">},</span>
    <span class="n">seed</span><span class="o">=</span><span class="n">seed</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">CodonVariantTable_h1</span><span class="o">.</span><span class="n">n_variants_df</span><span class="p">(</span><span class="n">samples</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>library</th>
      <th>sample</th>
      <th>count</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>lib_1</td>
      <td>barcoded variants</td>
      <td>25000</td>
    </tr>
    <tr>
      <th>1</th>
      <td>lib_2</td>
      <td>barcoded variants</td>
      <td>25000</td>
    </tr>
    <tr>
      <th>2</th>
      <td>all libraries</td>
      <td>barcoded variants</td>
      <td>50000</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>plot the number of variant support sequences as a histogram</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">p</span> <span class="o">=</span> <span class="n">CodonVariantTable_h1</span><span class="o">.</span><span class="n">plotVariantSupportHistogram</span><span class="p">(</span><span class="n">max_support</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
<span class="n">p</span> <span class="o">+=</span> <span class="n">theme_classic</span><span class="p">()</span> 
<span class="n">p</span> <span class="o">+=</span> <span class="n">theme</span><span class="p">(</span>
    <span class="n">panel_grid_major_x</span><span class="o">=</span><span class="n">element_blank</span><span class="p">(),</span>  <span class="c1"># no vertical grid lines</span>
    <span class="n">figure_size</span><span class="o">=</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">3</span><span class="p">)</span>
<span class="p">)</span>
<span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">(</span><span class="n">show</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/f7897c37c7a99975ceaeb21c2ae9cd13b03bfdda733ae51187c7dea945e095c9.png" src="_images/f7897c37c7a99975ceaeb21c2ae9cd13b03bfdda733ae51187c7dea945e095c9.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">CodonVariantTable_h2</span><span class="o">.</span><span class="n">n_variants_df</span><span class="p">(</span><span class="n">samples</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>library</th>
      <th>sample</th>
      <th>count</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>lib_1</td>
      <td>barcoded variants</td>
      <td>25000</td>
    </tr>
    <tr>
      <th>1</th>
      <td>lib_2</td>
      <td>barcoded variants</td>
      <td>25000</td>
    </tr>
    <tr>
      <th>2</th>
      <td>all libraries</td>
      <td>barcoded variants</td>
      <td>50000</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">p</span> <span class="o">=</span> <span class="n">CodonVariantTable_h2</span><span class="o">.</span><span class="n">plotVariantSupportHistogram</span><span class="p">(</span><span class="n">max_support</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
<span class="n">p</span> <span class="o">+=</span> <span class="n">theme_classic</span><span class="p">()</span> 
<span class="n">p</span> <span class="o">+=</span> <span class="n">theme</span><span class="p">(</span>
    <span class="n">panel_grid_major_x</span><span class="o">=</span><span class="n">element_blank</span><span class="p">(),</span>  <span class="c1"># no vertical grid lines</span>
    <span class="n">figure_size</span><span class="o">=</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">3</span><span class="p">)</span>
<span class="p">)</span>
<span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">(</span><span class="n">show</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/f7897c37c7a99975ceaeb21c2ae9cd13b03bfdda733ae51187c7dea945e095c9.png" src="_images/f7897c37c7a99975ceaeb21c2ae9cd13b03bfdda733ae51187c7dea945e095c9.png" />
</div>
</div>
</section>
<section id="variant-phenotypes-and-enrichments">
<h2>Variant phenotypes and enrichments<a class="headerlink" href="#variant-phenotypes-and-enrichments" title="Link to this heading">¶</a></h2>
<p>Now that we have simulated libraries of variants $v_i \in V$, we’ll assign their respective ground truth phenotypes and enrichments.</p>
<p>We’ll start by computing latent phenotype, $\phi(v_i)$. Recall that each of the two homologs has a <code class="docutils literal notranslate"><span class="pre">SigmoidPhenotypeSimulator</span></code> object that stores the latent effects of individual mutations. We will use these objects to compute the latent phenotypes of all variants in the libraries.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># we&#39;ll store a collection of functions for each homolog</span>
<span class="n">phenotype_fxn_dict_h1</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;latentPhenotype&quot;</span> <span class="p">:</span> <span class="n">SigmoidPhenotype_h1</span><span class="o">.</span><span class="n">latentPhenotype</span><span class="p">}</span>
<span class="n">phenotype_fxn_dict_h2</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;latentPhenotype&quot;</span> <span class="p">:</span> <span class="n">SigmoidPhenotype_h2</span><span class="o">.</span><span class="n">latentPhenotype</span><span class="p">}</span>

<span class="c1"># add the latent phenotype to the barcode variant table</span>
<span class="n">CodonVariantTable_h1</span><span class="o">.</span><span class="n">barcode_variant_df</span><span class="p">[</span><span class="s1">&#39;latent_phenotype&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">CodonVariantTable_h1</span><span class="o">.</span><span class="n">barcode_variant_df</span><span class="p">[</span><span class="s1">&#39;aa_substitutions&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
    <span class="n">phenotype_fxn_dict_h1</span><span class="p">[</span><span class="s2">&quot;latentPhenotype&quot;</span><span class="p">]</span>
<span class="p">)</span>
<span class="n">CodonVariantTable_h2</span><span class="o">.</span><span class="n">barcode_variant_df</span><span class="p">[</span><span class="s1">&#39;latent_phenotype&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">CodonVariantTable_h2</span><span class="o">.</span><span class="n">barcode_variant_df</span><span class="p">[</span><span class="s1">&#39;aa_substitutions&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
    <span class="n">phenotype_fxn_dict_h2</span><span class="p">[</span><span class="s2">&quot;latentPhenotype&quot;</span><span class="p">]</span>
<span class="p">)</span>

<span class="c1"># print a sample of the variant table</span>
<span class="n">CodonVariantTable_h1</span><span class="o">.</span><span class="n">barcode_variant_df</span><span class="p">[[</span><span class="s2">&quot;library&quot;</span><span class="p">,</span> <span class="s2">&quot;aa_substitutions&quot;</span><span class="p">,</span> <span class="s2">&quot;latent_phenotype&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>library</th>
      <th>aa_substitutions</th>
      <th>latent_phenotype</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>lib_1</td>
      <td>V24Q S25R H27P D29L H33C S41L C44F S47F</td>
      <td>-11.096139</td>
    </tr>
    <tr>
      <th>1</th>
      <td>lib_1</td>
      <td>V8P H27T P30* L49P</td>
      <td>-11.106200</td>
    </tr>
    <tr>
      <th>2</th>
      <td>lib_1</td>
      <td>V19P V43Q</td>
      <td>-2.878203</td>
    </tr>
    <tr>
      <th>3</th>
      <td>lib_1</td>
      <td>R20Y</td>
      <td>5.161623</td>
    </tr>
    <tr>
      <th>4</th>
      <td>lib_1</td>
      <td></td>
      <td>5.000000</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>Next, we can compute an <em>observed phenotype</em>, $p(v, d)$ = $g_{\theta}(\phi(v, d))$, where $g_{\theta}$ is the global epistasis function implimented as a flexible sigmoid with two free parameters, $\theta_{s}$ and $\theta_{b}$, that serve as a scale and bias, respectively. Concretely, $g_{\theta}(z) = \theta_{b} + \frac{\theta_{s}}{1 + e^{-z}}$. For the purposes of this analysis, we compute a fixed value for the $\theta_{b}$ parameter, based upon the specified parameters for $\theta_{s}$ (<code class="docutils literal notranslate"><span class="pre">sigmoid_phenotype_scale</span></code>) and $\phi(v_{WT})$ (<code class="docutils literal notranslate"><span class="pre">wt_latent</span></code>) such that the reference wildtype phenotype is exactly 0.</p>
<p>Note that while the <code class="docutils literal notranslate"><span class="pre">SigmoidPhenotypeSimulator</span></code> object (as the name suggests) provides its own global epistasis object for computing phenotype and enrichments, we will only be using it to get mutational effects and the associated latent phenotypes. For computing the post-latent phenotypes and enrichments, we will use the <code class="docutils literal notranslate"><span class="pre">multidms</span></code> biophysical model such that we’re simulating under the same model we’ll be fitting as described above.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># define numpy wrapper for multidms native jax sigmoid fxn</span>
<span class="k">def</span> <span class="nf">g</span><span class="p">(</span><span class="n">z</span><span class="p">:</span><span class="nb">float</span><span class="p">,</span> <span class="n">ge_scale</span><span class="o">=</span><span class="n">sigmoid_phenotype_scale</span><span class="p">,</span> <span class="n">wt_latent</span><span class="o">=</span><span class="n">wt_latent</span><span class="p">):</span>

    <span class="c1"># ensure the reference phenotype is 0</span>
    <span class="n">ge_bias</span> <span class="o">=</span> <span class="o">-</span><span class="n">sigmoid_phenotype_scale</span> <span class="o">/</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">exp</span><span class="p">(</span><span class="o">-</span><span class="n">wt_latent</span><span class="p">))</span>
    <span class="k">return</span> <span class="n">ge_bias</span> <span class="o">+</span> <span class="p">(</span><span class="n">ge_scale</span><span class="o">/</span><span class="p">(</span><span class="mi">1</span><span class="o">+</span><span class="n">np</span><span class="o">.</span><span class="n">exp</span><span class="p">(</span><span class="o">-</span><span class="n">z</span><span class="p">)))</span>
</pre></div>
</div>
</div>
</div>
<p>Visualize the sigmoidal function that we’ll be using to map latent phenotypes to observed phenotypes. Place vertical line at the reference (“h1”) wildtype latent phenotype, as well as a dashed line for the non-reference (“h2”) phenotype:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">resolution</span> <span class="o">=</span> <span class="p">(</span><span class="o">-</span><span class="mi">10</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">100</span><span class="p">)</span>
<span class="n">p</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">ggplot</span><span class="p">(</span>
        <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
            <span class="p">{</span>
                <span class="s2">&quot;x&quot;</span><span class="p">:</span><span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="o">*</span><span class="n">resolution</span><span class="p">),</span> 
                <span class="s2">&quot;y&quot;</span><span class="p">:</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="n">g</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="o">*</span><span class="n">resolution</span><span class="p">))))</span>
            <span class="p">}</span>
        <span class="p">),</span> 
    <span class="p">)</span>
    <span class="o">+</span> <span class="n">geom_line</span><span class="p">(</span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="s2">&quot;x&quot;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s2">&quot;y&quot;</span><span class="p">))</span>
    <span class="o">+</span> <span class="n">geom_vline</span><span class="p">(</span><span class="n">xintercept</span><span class="o">=</span><span class="n">SigmoidPhenotype_h1</span><span class="o">.</span><span class="n">wt_latent</span><span class="p">)</span>
    <span class="o">+</span> <span class="n">geom_vline</span><span class="p">(</span><span class="n">xintercept</span><span class="o">=</span><span class="n">SigmoidPhenotype_h2</span><span class="o">.</span><span class="n">wt_latent</span><span class="p">,</span> <span class="n">linetype</span><span class="o">=</span><span class="s2">&quot;dotted&quot;</span><span class="p">)</span>
    <span class="o">+</span> <span class="n">labs</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="s2">&quot;latent phenotype&quot;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s2">&quot;observed phenotype&quot;</span><span class="p">)</span>
    <span class="o">+</span> <span class="n">theme_classic</span><span class="p">()</span>
    <span class="o">+</span> <span class="n">theme</span><span class="p">(</span><span class="n">figure_size</span><span class="o">=</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">))</span>
<span class="p">)</span>
<span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">(</span><span class="n">show</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/37cc0789c6fb6557dbc40f7fdcdb63c8cd1dd4cc0db83e79460978f10e71f903.png" src="_images/37cc0789c6fb6557dbc40f7fdcdb63c8cd1dd4cc0db83e79460978f10e71f903.png" />
</div>
</div>
<p>Next, assign each barcoded variant an <em>observed phenotype</em>, $p(v, d)$, as well as an <em>observed enrichment</em> , $2^{p(v, d)}$.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">subs</span> <span class="o">=</span> <span class="n">CodonVariantTable_h1</span><span class="o">.</span><span class="n">barcode_variant_df</span><span class="p">[</span><span class="s1">&#39;aa_substitutions&#39;</span><span class="p">]</span>
<span class="n">phenotype_fxn_dict_h1</span><span class="p">[</span><span class="s2">&quot;observedPhenotype&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">g</span><span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">phenotype_fxn_dict_h1</span><span class="p">[</span><span class="s2">&quot;latentPhenotype&quot;</span><span class="p">](</span><span class="n">x</span><span class="p">)))</span>
<span class="n">CodonVariantTable_h1</span><span class="o">.</span><span class="n">barcode_variant_df</span><span class="p">[</span><span class="s1">&#39;observed_phenotype&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">subs</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
    <span class="n">phenotype_fxn_dict_h1</span><span class="p">[</span><span class="s2">&quot;observedPhenotype&quot;</span><span class="p">]</span>
<span class="p">)</span>
<span class="n">phenotype_fxn_dict_h1</span><span class="p">[</span><span class="s2">&quot;observedEnrichment&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="mi">2</span> <span class="o">**</span> <span class="p">(</span><span class="n">phenotype_fxn_dict_h1</span><span class="p">[</span><span class="s2">&quot;observedPhenotype&quot;</span><span class="p">](</span><span class="n">x</span><span class="p">))</span>
<span class="n">CodonVariantTable_h1</span><span class="o">.</span><span class="n">barcode_variant_df</span><span class="p">[</span><span class="s1">&#39;observed_enrichment&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">subs</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
    <span class="n">phenotype_fxn_dict_h1</span><span class="p">[</span><span class="s2">&quot;observedEnrichment&quot;</span><span class="p">]</span>
<span class="p">)</span>

<span class="n">subs</span> <span class="o">=</span> <span class="n">CodonVariantTable_h2</span><span class="o">.</span><span class="n">barcode_variant_df</span><span class="p">[</span><span class="s1">&#39;aa_substitutions&#39;</span><span class="p">]</span>
<span class="n">phenotype_fxn_dict_h2</span><span class="p">[</span><span class="s2">&quot;observedPhenotype&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">g</span><span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">phenotype_fxn_dict_h2</span><span class="p">[</span><span class="s2">&quot;latentPhenotype&quot;</span><span class="p">](</span><span class="n">x</span><span class="p">)))</span>
<span class="n">CodonVariantTable_h2</span><span class="o">.</span><span class="n">barcode_variant_df</span><span class="p">[</span><span class="s1">&#39;observed_phenotype&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">subs</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
    <span class="n">phenotype_fxn_dict_h2</span><span class="p">[</span><span class="s2">&quot;observedPhenotype&quot;</span><span class="p">]</span>
<span class="p">)</span>
<span class="n">phenotype_fxn_dict_h2</span><span class="p">[</span><span class="s2">&quot;observedEnrichment&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="mi">2</span> <span class="o">**</span> <span class="p">(</span><span class="n">phenotype_fxn_dict_h2</span><span class="p">[</span><span class="s2">&quot;observedPhenotype&quot;</span><span class="p">](</span><span class="n">x</span><span class="p">))</span>
<span class="n">CodonVariantTable_h2</span><span class="o">.</span><span class="n">barcode_variant_df</span><span class="p">[</span><span class="s1">&#39;observed_enrichment&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">subs</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
    <span class="n">phenotype_fxn_dict_h2</span><span class="p">[</span><span class="s2">&quot;observedEnrichment&quot;</span><span class="p">]</span>
<span class="p">)</span>
<span class="n">variants_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span>
    <span class="p">[</span>
        <span class="n">CodonVariantTable_h1</span><span class="o">.</span><span class="n">barcode_variant_df</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">homolog</span><span class="o">=</span><span class="s2">&quot;h1&quot;</span><span class="p">),</span> 
        <span class="n">CodonVariantTable_h2</span><span class="o">.</span><span class="n">barcode_variant_df</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">homolog</span><span class="o">=</span><span class="s2">&quot;h2&quot;</span><span class="p">)</span>
    <span class="p">]</span>
<span class="p">)</span>

<span class="n">variants_df</span><span class="p">[[</span><span class="s2">&quot;aa_substitutions&quot;</span><span class="p">,</span> <span class="s2">&quot;latent_phenotype&quot;</span><span class="p">,</span> <span class="s2">&quot;observed_phenotype&quot;</span><span class="p">,</span> <span class="s2">&quot;observed_enrichment&quot;</span><span class="p">,</span> <span class="s2">&quot;homolog&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>aa_substitutions</th>
      <th>latent_phenotype</th>
      <th>observed_phenotype</th>
      <th>observed_enrichment</th>
      <th>homolog</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>V24Q S25R H27P D29L H33C S41L C44F S47F</td>
      <td>-11.096139</td>
      <td>-5.959752</td>
      <td>0.016067</td>
      <td>h1</td>
    </tr>
    <tr>
      <th>1</th>
      <td>V8P H27T P30* L49P</td>
      <td>-11.106200</td>
      <td>-5.959753</td>
      <td>0.016067</td>
      <td>h1</td>
    </tr>
    <tr>
      <th>2</th>
      <td>V19P V43Q</td>
      <td>-2.878203</td>
      <td>-5.640393</td>
      <td>0.020048</td>
      <td>h1</td>
    </tr>
    <tr>
      <th>3</th>
      <td>R20Y</td>
      <td>5.161623</td>
      <td>0.005959</td>
      <td>1.004139</td>
      <td>h1</td>
    </tr>
    <tr>
      <th>4</th>
      <td></td>
      <td>5.000000</td>
      <td>0.000000</td>
      <td>1.000000</td>
      <td>h1</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>Note that The enrichment score transforms phenotypes to be between 0 and 1, and can thus be used to compute the number of counts for each variant in the post-selection library using a multinomial distribution where the probability of observing a variant is proportional to its enrichment. Below, we plot each variant’s latent phenotype against its observed enrichment:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">p</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">ggplot</span><span class="p">(</span>
        <span class="n">variants_df</span><span class="p">,</span> 
        <span class="n">aes</span><span class="p">(</span>
            <span class="n">x</span> <span class="o">=</span> <span class="s2">&quot;latent_phenotype&quot;</span><span class="p">,</span>
            <span class="n">y</span> <span class="o">=</span> <span class="s2">&quot;observed_enrichment&quot;</span>
        <span class="p">)</span>
    <span class="p">)</span>
    <span class="o">+</span> <span class="n">geom_point</span><span class="p">()</span>
    <span class="o">+</span> <span class="n">geom_vline</span><span class="p">(</span><span class="n">xintercept</span><span class="o">=</span><span class="n">SigmoidPhenotype_h1</span><span class="o">.</span><span class="n">wt_latent</span><span class="p">)</span>
    <span class="o">+</span> <span class="n">geom_vline</span><span class="p">(</span><span class="n">xintercept</span><span class="o">=</span><span class="n">SigmoidPhenotype_h2</span><span class="o">.</span><span class="n">wt_latent</span><span class="p">,</span> <span class="n">linetype</span><span class="o">=</span><span class="s2">&quot;dotted&quot;</span><span class="p">)</span>
    <span class="o">+</span> <span class="n">theme_classic</span><span class="p">()</span>
    <span class="o">+</span> <span class="n">theme</span><span class="p">(</span>
        <span class="n">figure_size</span><span class="o">=</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">3</span><span class="p">),</span>
        <span class="n">axis_text_x</span><span class="o">=</span><span class="n">element_text</span><span class="p">(</span><span class="n">angle</span><span class="o">=</span><span class="mi">90</span><span class="p">),</span>
        <span class="n">panel_grid_major_x</span><span class="o">=</span><span class="n">element_blank</span><span class="p">(),</span>  <span class="c1"># no vertical grid lines</span>
    <span class="p">)</span>
    <span class="o">+</span> <span class="n">labs</span><span class="p">(</span>
        <span class="n">x</span><span class="o">=</span><span class="s2">&quot;$\phi$&quot;</span><span class="p">,</span>
        <span class="n">y</span><span class="o">=</span><span class="s2">&quot;observed enrichment&quot;</span>
    <span class="p">)</span>
    <span class="c1"># zoom in on the x axis limits</span>
    <span class="o">+</span> <span class="n">scale_x_continuous</span><span class="p">(</span><span class="n">limits</span><span class="o">=</span><span class="p">(</span><span class="o">-</span><span class="mi">10</span><span class="p">,</span> <span class="mi">10</span><span class="p">))</span>
<span class="p">)</span>
<span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">(</span><span class="n">show</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/712448d9fd582981de0e5019b2b4cb9db9742008690b5ef29302819f4c193ffe.png" src="_images/712448d9fd582981de0e5019b2b4cb9db9742008690b5ef29302819f4c193ffe.png" />
</div>
</div>
</section>
<section id="pre-and-post-selection-variant-read-counts">
<h2>Pre and post-selection variant read counts<a class="headerlink" href="#pre-and-post-selection-variant-read-counts" title="Link to this heading">¶</a></h2>
<p>Next, we will simulate both pre and post-selection counts of each variant subjected to an experimental selection step. Post-selection counts will be simulated by applying the specified selection bottleneck(s) to the pre-selection counts. To do this, we will use the functionality provided by <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.simulate.html#dms_variants.simulate.simulateSampleCounts">dms_variants.simulate.simulateSampleCounts</a> and feed it our custom phenotype function.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">bottlenecks</span> <span class="o">=</span> <span class="p">{</span>
    <span class="n">name</span> <span class="p">:</span> <span class="n">variants_per_lib</span> <span class="o">*</span> <span class="n">scalar</span>
    <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">scalar</span> <span class="ow">in</span> <span class="n">bottleneck_variants_per_lib_scalar</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
<span class="p">}</span>

<span class="c1"># do this independently for each of the homologs.</span>
<span class="n">counts_h1</span><span class="p">,</span> <span class="n">counts_h2</span> <span class="o">=</span> <span class="p">[</span>
    <span class="n">dms_variants</span><span class="o">.</span><span class="n">simulate</span><span class="o">.</span><span class="n">simulateSampleCounts</span><span class="p">(</span>
        <span class="n">variants</span><span class="o">=</span><span class="n">variants</span><span class="p">,</span>
        <span class="n">phenotype_func</span><span class="o">=</span><span class="n">pheno_fxn_dict</span><span class="p">[</span><span class="s2">&quot;observedEnrichment&quot;</span><span class="p">],</span>
        <span class="n">variant_error_rate</span><span class="o">=</span><span class="n">variant_error_rate</span><span class="p">,</span>
        <span class="n">pre_sample</span><span class="o">=</span><span class="p">{</span>
            <span class="s2">&quot;total_count&quot;</span><span class="p">:</span> <span class="n">variants_per_lib</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">poisson</span><span class="p">(</span><span class="n">avgdepth_per_variant</span><span class="p">),</span>
            <span class="s2">&quot;uniformity&quot;</span><span class="p">:</span> <span class="n">lib_uniformity</span><span class="p">,</span>
        <span class="p">},</span>
        <span class="n">pre_sample_name</span><span class="o">=</span><span class="s2">&quot;pre-selection&quot;</span><span class="p">,</span>
        <span class="n">post_samples</span><span class="o">=</span><span class="p">{</span>
            <span class="n">name</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;noise&quot;</span><span class="p">:</span> <span class="n">noise</span><span class="p">,</span>
                <span class="s2">&quot;total_count&quot;</span><span class="p">:</span> <span class="n">variants_per_lib</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">poisson</span><span class="p">(</span><span class="n">avgdepth_per_variant</span><span class="p">),</span>
                <span class="s2">&quot;bottleneck&quot;</span><span class="p">:</span> <span class="n">bottleneck</span><span class="p">,</span>
            <span class="p">}</span>
            <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">bottleneck</span> <span class="ow">in</span> <span class="n">bottlenecks</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
        <span class="p">},</span>
        <span class="n">seed</span><span class="o">=</span><span class="n">seed</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="k">for</span> <span class="n">variants</span><span class="p">,</span> <span class="n">pheno_fxn_dict</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span>
        <span class="p">[</span><span class="n">CodonVariantTable_h1</span><span class="p">,</span> <span class="n">CodonVariantTable_h2</span><span class="p">],</span> <span class="p">[</span><span class="n">phenotype_fxn_dict_h1</span><span class="p">,</span> <span class="n">phenotype_fxn_dict_h2</span><span class="p">]</span>
    <span class="p">)</span>
<span class="p">]</span>
<span class="n">CodonVariantTable_h1</span><span class="o">.</span><span class="n">add_sample_counts_df</span><span class="p">(</span><span class="n">counts_h1</span><span class="p">)</span>
<span class="n">CodonVariantTable_h2</span><span class="o">.</span><span class="n">add_sample_counts_df</span><span class="p">(</span><span class="n">counts_h2</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Plot the number of counts for each variant in each sample. The horizontal dashed line shows the total number of variants. The plot shows that all variants are well-sampled in the pre-selection libraries, but that post- selection some variants are sampled more or less. This is expected since selection will decrease and increase the frequency of variants:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">variants</span><span class="p">,</span> <span class="n">title</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">([</span><span class="n">CodonVariantTable_h1</span><span class="p">,</span> <span class="n">CodonVariantTable_h2</span><span class="p">],</span> <span class="p">[</span><span class="s1">&#39;homolog 1&#39;</span><span class="p">,</span> <span class="s1">&#39;homolog 2&#39;</span><span class="p">]):</span>
    <span class="n">p</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">plotCumulVariantCounts</span><span class="p">()</span>
    <span class="n">p</span> <span class="o">+=</span> <span class="n">theme_classic</span><span class="p">()</span> 
    <span class="n">p</span> <span class="o">+=</span> <span class="n">theme</span><span class="p">(</span>
        <span class="n">panel_grid_major_x</span><span class="o">=</span><span class="n">element_blank</span><span class="p">(),</span>  <span class="c1"># no vertical grid lines</span>
        <span class="n">figure_size</span><span class="o">=</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">5</span><span class="p">)</span>
    <span class="p">)</span>
    <span class="n">p</span> <span class="o">+=</span> <span class="n">labs</span><span class="p">(</span><span class="n">title</span><span class="o">=</span><span class="n">title</span><span class="p">)</span>
    <span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">(</span><span class="n">show</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/f76bc2b39ae7026a25aa8d297c6c69fc3a4af4f0d97df3ddb430752917ccc615.png" src="_images/f76bc2b39ae7026a25aa8d297c6c69fc3a4af4f0d97df3ddb430752917ccc615.png" />
<img alt="_images/495bf43be92a02321bf5d44ccbc62d5e5f70c82ea4d26ee602407845fe3570a0.png" src="_images/495bf43be92a02321bf5d44ccbc62d5e5f70c82ea4d26ee602407845fe3570a0.png" />
</div>
</div>
<p>Distribution of the number of amino-acid mutations per variant in each sample. As expected, mutations go down after selection:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">variants</span><span class="p">,</span> <span class="n">title</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">([</span><span class="n">CodonVariantTable_h1</span><span class="p">,</span> <span class="n">CodonVariantTable_h2</span><span class="p">],</span> <span class="p">[</span><span class="s1">&#39;homolog 1&#39;</span><span class="p">,</span> <span class="s1">&#39;homolog 2&#39;</span><span class="p">]):</span>
    <span class="n">p</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">plotNumMutsHistogram</span><span class="p">(</span><span class="n">mut_type</span><span class="o">=</span><span class="s2">&quot;aa&quot;</span><span class="p">)</span>
    <span class="n">p</span> <span class="o">+=</span> <span class="n">theme_classic</span><span class="p">()</span> 
    <span class="n">p</span> <span class="o">+=</span> <span class="n">theme</span><span class="p">(</span><span class="n">figure_size</span><span class="o">=</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">5</span><span class="p">))</span>
    <span class="n">p</span> <span class="o">+=</span> <span class="n">labs</span><span class="p">(</span><span class="n">title</span><span class="o">=</span><span class="n">title</span><span class="p">)</span>
    <span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">(</span><span class="n">show</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/6896826f3b922af22a8dee1438b90abd30ec9316fc19ef512049f885ad067cfb.png" src="_images/6896826f3b922af22a8dee1438b90abd30ec9316fc19ef512049f885ad067cfb.png" />
<img alt="_images/47e64ad6d475e6ab9d42d2048e10f84ff7b9fe5dab1bd61cc5a71fc844102714.png" src="_images/47e64ad6d475e6ab9d42d2048e10f84ff7b9fe5dab1bd61cc5a71fc844102714.png" />
</div>
</div>
<p>Plot how thoroughly amino-acid mutations are sampled. The plots below show that the stop mutations are sampled very poorly post-selection because they are eliminated during selection:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">variants</span><span class="p">,</span> <span class="n">title</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">([</span><span class="n">CodonVariantTable_h1</span><span class="p">,</span> <span class="n">CodonVariantTable_h2</span><span class="p">],</span> <span class="p">[</span><span class="s1">&#39;homolog 1&#39;</span><span class="p">,</span> <span class="s1">&#39;homolog 2&#39;</span><span class="p">]):</span>
    <span class="n">p</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">plotCumulMutCoverage</span><span class="p">(</span><span class="n">variant_type</span><span class="o">=</span><span class="s2">&quot;all&quot;</span><span class="p">,</span> <span class="n">mut_type</span><span class="o">=</span><span class="s2">&quot;aa&quot;</span><span class="p">)</span>
    <span class="n">p</span> <span class="o">+=</span> <span class="n">theme_classic</span><span class="p">()</span> 
    <span class="n">p</span> <span class="o">+=</span> <span class="n">theme</span><span class="p">(</span><span class="n">figure_size</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="mi">5</span><span class="p">))</span>
    <span class="n">p</span> <span class="o">+=</span> <span class="n">theme</span><span class="p">(</span><span class="n">axis_text_x</span><span class="o">=</span><span class="n">element_text</span><span class="p">(</span><span class="n">angle</span><span class="o">=</span><span class="mi">90</span><span class="p">))</span>
    <span class="n">p</span> <span class="o">+=</span> <span class="n">labs</span><span class="p">(</span><span class="n">title</span><span class="o">=</span><span class="n">title</span><span class="p">)</span>
    <span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">(</span><span class="n">show</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/894b5a1170a88ff42a67340eff3087dfb8a28f7e4b991d91a4f021a92afc27eb.png" src="_images/894b5a1170a88ff42a67340eff3087dfb8a28f7e4b991d91a4f021a92afc27eb.png" />
<img alt="_images/f460e4828c7975121abee6266515c8ab5c3928541505584fe47d4ae1f9b9266b.png" src="_images/f460e4828c7975121abee6266515c8ab5c3928541505584fe47d4ae1f9b9266b.png" />
</div>
</div>
</section>
<section id="prep-training-data">
<h2>Prep training data<a class="headerlink" href="#prep-training-data" title="Link to this heading">¶</a></h2>
<p>Prepare the training dataframes for fitting our <code class="docutils literal notranslate"><span class="pre">multidms</span></code> models.</p>
<p>As this is a joint-fitting approach, we combine homolog variants from each of the corresponding library replicates into a single training dataset. Additionally, for each replicate dataset, we’ll train models on each of the following targets representing different levels of noise in the data:</p>
<ol class="arabic simple">
<li><p>Ground truth <em>observed phenotype</em> - this target acts as a control and benchmarks the ability of the model to recover the true latent effects and shifts controlling for other sources of noise.</p></li>
<li><p><em>Loose bottleneck</em> counts derived functional scores - these scores are derived from the observed enrichments, and used to asses model performance in the context of realistic experimental noise.</p></li>
<li><p><em>Tight bottleneck</em> counts derived functional scores - the same as above, but with a tighter bottleneck for more extreme cases of noise.</p></li>
</ol>
<p>For further details on how to prepare data and fit models, see the <code class="docutils literal notranslate"><span class="pre">multidms</span></code> <a class="reference external" href="https://matsengrp.github.io/multidms/fit_delta_BA1_example.html">quick-start tutorial</a>.</p>
<p>Start by creating training data with ground truth phenotype target. Because the barcode replicates share ground truth phenotypes we can can collapse the counts accross replicates by simple dropping duplicates.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">req_cols</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;library&quot;</span><span class="p">,</span> <span class="s2">&quot;homolog&quot;</span><span class="p">,</span> <span class="s2">&quot;aa_substitutions&quot;</span><span class="p">,</span> <span class="s2">&quot;func_score_type&quot;</span><span class="p">,</span> <span class="s2">&quot;func_score&quot;</span><span class="p">]</span>
<span class="n">ground_truth_training_set</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span>
        <span class="p">[</span>
            <span class="n">variants</span><span class="o">.</span><span class="n">barcode_variant_df</span><span class="p">[[</span><span class="s2">&quot;library&quot;</span><span class="p">,</span> <span class="s2">&quot;aa_substitutions&quot;</span><span class="p">,</span> <span class="s2">&quot;observed_phenotype&quot;</span><span class="p">,</span> <span class="s2">&quot;latent_phenotype&quot;</span><span class="p">]]</span>
            <span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span>
            <span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">homolog</span><span class="o">=</span><span class="n">homolog</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">variants</span><span class="p">,</span> <span class="n">homolog</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">([</span><span class="n">CodonVariantTable_h1</span><span class="p">,</span> <span class="n">CodonVariantTable_h2</span><span class="p">],</span> <span class="p">[</span><span class="s1">&#39;h1&#39;</span><span class="p">,</span> <span class="s1">&#39;h2&#39;</span><span class="p">])</span>
        <span class="p">]</span>
    <span class="p">)</span>
    <span class="o">.</span><span class="n">melt</span><span class="p">(</span>
        <span class="n">id_vars</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;library&quot;</span><span class="p">,</span> <span class="s2">&quot;aa_substitutions&quot;</span><span class="p">,</span> <span class="s2">&quot;homolog&quot;</span><span class="p">,</span> <span class="s2">&quot;latent_phenotype&quot;</span><span class="p">],</span>
        <span class="n">value_vars</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;observed_phenotype&quot;</span><span class="p">],</span>
        <span class="n">var_name</span><span class="o">=</span><span class="s2">&quot;func_score_type&quot;</span><span class="p">,</span>
        <span class="n">value_name</span><span class="o">=</span><span class="s2">&quot;func_score&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="p">[</span><span class="n">req_cols</span><span class="p">]</span>
    <span class="o">.</span><span class="n">astype</span><span class="p">({</span><span class="n">c</span><span class="p">:</span><span class="nb">str</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">req_cols</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">]})</span>
<span class="p">)</span>
<span class="n">ground_truth_training_set</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>library</th>
      <th>homolog</th>
      <th>aa_substitutions</th>
      <th>func_score_type</th>
      <th>func_score</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>lib_1</td>
      <td>h1</td>
      <td>V24Q S25R H27P D29L H33C S41L C44F S47F</td>
      <td>observed_phenotype</td>
      <td>-5.96</td>
    </tr>
    <tr>
      <th>1</th>
      <td>lib_1</td>
      <td>h1</td>
      <td>V8P H27T P30* L49P</td>
      <td>observed_phenotype</td>
      <td>-5.96</td>
    </tr>
    <tr>
      <th>2</th>
      <td>lib_1</td>
      <td>h1</td>
      <td>V19P V43Q</td>
      <td>observed_phenotype</td>
      <td>-5.64</td>
    </tr>
    <tr>
      <th>3</th>
      <td>lib_1</td>
      <td>h1</td>
      <td>R20Y</td>
      <td>observed_phenotype</td>
      <td>0.01</td>
    </tr>
    <tr>
      <th>4</th>
      <td>lib_1</td>
      <td>h1</td>
      <td></td>
      <td>observed_phenotype</td>
      <td>0.00</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>Next, compute functional scores from pre-post counts in each bottleneck <em>after</em> aggregating the barcode replicate counts for unique variants. The <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.codonvarianttable.html#dms_variants.codonvarianttable.CodonVariantTable.func_scores">dms_variants.CodonVariantTable.func_scores</a> method provides the ability to compute functional scores from pre and post-selection counts. We’ll use this to compute functional scores for each of the three targets:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">bottle_cbf</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span>
    <span class="p">[</span>
        <span class="p">(</span>
            <span class="n">variants</span>
            <span class="o">.</span><span class="n">func_scores</span><span class="p">(</span><span class="s2">&quot;pre-selection&quot;</span><span class="p">,</span> <span class="n">by</span><span class="o">=</span><span class="s2">&quot;aa_substitutions&quot;</span><span class="p">,</span> <span class="n">libraries</span><span class="o">=</span><span class="n">libs</span><span class="p">,</span> <span class="n">syn_as_wt</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">homolog</span><span class="o">=</span><span class="n">homolog</span><span class="p">)</span>
            <span class="o">.</span><span class="n">rename</span><span class="p">({</span><span class="s2">&quot;post_sample&quot;</span><span class="p">:</span><span class="s2">&quot;func_score_type&quot;</span><span class="p">},</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
            <span class="p">[</span><span class="n">req_cols</span><span class="p">]</span>
            <span class="o">.</span><span class="n">astype</span><span class="p">({</span><span class="n">c</span><span class="p">:</span><span class="nb">str</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">req_cols</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">]})</span>
        <span class="p">)</span>
        <span class="k">for</span> <span class="n">variants</span><span class="p">,</span> <span class="n">homolog</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">([</span><span class="n">CodonVariantTable_h1</span><span class="p">,</span> <span class="n">CodonVariantTable_h2</span><span class="p">],</span> <span class="p">[</span><span class="s1">&#39;h1&#39;</span><span class="p">,</span> <span class="s1">&#39;h2&#39;</span><span class="p">])</span>
    <span class="p">]</span>
<span class="p">)</span>
<span class="n">bottle_cbf</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>library</th>
      <th>homolog</th>
      <th>aa_substitutions</th>
      <th>func_score_type</th>
      <th>func_score</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>lib_1</td>
      <td>h1</td>
      <td>Q28P</td>
      <td>loose_bottle</td>
      <td>-0.021369</td>
    </tr>
    <tr>
      <th>1</th>
      <td>lib_1</td>
      <td>h1</td>
      <td>G50E</td>
      <td>loose_bottle</td>
      <td>-0.067491</td>
    </tr>
    <tr>
      <th>2</th>
      <td>lib_1</td>
      <td>h1</td>
      <td>R48G</td>
      <td>loose_bottle</td>
      <td>-1.291522</td>
    </tr>
    <tr>
      <th>3</th>
      <td>lib_1</td>
      <td>h1</td>
      <td>S3R G10Y L49H</td>
      <td>loose_bottle</td>
      <td>-5.759732</td>
    </tr>
    <tr>
      <th>4</th>
      <td>lib_1</td>
      <td>h1</td>
      <td>F18T Q21E C44T</td>
      <td>loose_bottle</td>
      <td>-5.313711</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>Finally, combine the two dataframes computed above and classify the variants based on the number of amino acid substitutions.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">classify_variant</span><span class="p">(</span><span class="n">aa_subs</span><span class="p">):</span>
    <span class="k">if</span> <span class="s2">&quot;*&quot;</span> <span class="ow">in</span> <span class="n">aa_subs</span><span class="p">:</span>
        <span class="k">return</span> <span class="s2">&quot;stop&quot;</span>
    <span class="k">elif</span> <span class="n">aa_subs</span> <span class="o">==</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
        <span class="k">return</span> <span class="s2">&quot;wildtype&quot;</span>
    <span class="k">elif</span> <span class="nb">len</span><span class="p">(</span><span class="n">aa_subs</span><span class="o">.</span><span class="n">split</span><span class="p">())</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
        <span class="k">return</span> <span class="s2">&quot;1 nonsynonymous&quot;</span>
    <span class="k">elif</span> <span class="nb">len</span><span class="p">(</span><span class="n">aa_subs</span><span class="o">.</span><span class="n">split</span><span class="p">())</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
        <span class="k">return</span> <span class="s2">&quot;&gt;1 nonsynonymous&quot;</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;unexpected aa_subs: </span><span class="si">{</span><span class="n">aa_subs</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="n">func_scores</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">ground_truth_training_set</span><span class="p">,</span> <span class="n">bottle_cbf</span><span class="p">])</span>
    <span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">variant_class</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;aa_substitutions&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">classify_variant</span><span class="p">))</span>
    <span class="o">.</span><span class="n">merge</span><span class="p">(</span>
        <span class="n">variants_df</span><span class="p">[[</span><span class="s2">&quot;aa_substitutions&quot;</span><span class="p">,</span> <span class="s2">&quot;homolog&quot;</span><span class="p">,</span> <span class="s2">&quot;latent_phenotype&quot;</span><span class="p">,</span> <span class="s2">&quot;library&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">(),</span>
        <span class="n">on</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;aa_substitutions&quot;</span><span class="p">,</span> <span class="s2">&quot;homolog&quot;</span><span class="p">,</span> <span class="s2">&quot;library&quot;</span><span class="p">],</span>
        <span class="n">how</span><span class="o">=</span><span class="s2">&quot;inner&quot;</span>
    <span class="p">)</span>
<span class="p">)</span>

<span class="c1"># categorical for plotting</span>
<span class="n">func_scores</span><span class="p">[</span><span class="s2">&quot;func_score_type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Categorical</span><span class="p">(</span>
    <span class="n">func_scores</span><span class="p">[</span><span class="s2">&quot;func_score_type&quot;</span><span class="p">],</span>
    <span class="n">categories</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;observed_phenotype&quot;</span><span class="p">,</span> <span class="s2">&quot;loose_bottle&quot;</span><span class="p">,</span> <span class="s2">&quot;tight_bottle&quot;</span><span class="p">],</span>
    <span class="n">ordered</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>

<span class="n">func_scores</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>library</th>
      <th>homolog</th>
      <th>aa_substitutions</th>
      <th>func_score_type</th>
      <th>func_score</th>
      <th>variant_class</th>
      <th>latent_phenotype</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>lib_1</td>
      <td>h1</td>
      <td>V24Q S25R H27P D29L H33C S41L C44F S47F</td>
      <td>observed_phenotype</td>
      <td>-5.959752</td>
      <td>&gt;1 nonsynonymous</td>
      <td>-11.096139</td>
    </tr>
    <tr>
      <th>1</th>
      <td>lib_1</td>
      <td>h1</td>
      <td>V8P H27T P30* L49P</td>
      <td>observed_phenotype</td>
      <td>-5.959753</td>
      <td>stop</td>
      <td>-11.106200</td>
    </tr>
    <tr>
      <th>2</th>
      <td>lib_1</td>
      <td>h1</td>
      <td>V19P V43Q</td>
      <td>observed_phenotype</td>
      <td>-5.640393</td>
      <td>&gt;1 nonsynonymous</td>
      <td>-2.878203</td>
    </tr>
    <tr>
      <th>3</th>
      <td>lib_1</td>
      <td>h1</td>
      <td>R20Y</td>
      <td>observed_phenotype</td>
      <td>0.005959</td>
      <td>1 nonsynonymous</td>
      <td>5.161623</td>
    </tr>
    <tr>
      <th>4</th>
      <td>lib_1</td>
      <td>h1</td>
      <td></td>
      <td>observed_phenotype</td>
      <td>0.000000</td>
      <td>wildtype</td>
      <td>5.000000</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>181363</th>
      <td>lib_2</td>
      <td>h2</td>
      <td>E9L Q21L V24Q S25V D29V D38Q</td>
      <td>tight_bottle</td>
      <td>-6.184609</td>
      <td>&gt;1 nonsynonymous</td>
      <td>-11.209818</td>
    </tr>
    <tr>
      <th>181364</th>
      <td>lib_2</td>
      <td>h2</td>
      <td>S47T V49E</td>
      <td>tight_bottle</td>
      <td>1.125946</td>
      <td>&gt;1 nonsynonymous</td>
      <td>3.861711</td>
    </tr>
    <tr>
      <th>181365</th>
      <td>lib_2</td>
      <td>h2</td>
      <td>G10P I35S G50L</td>
      <td>tight_bottle</td>
      <td>-3.634411</td>
      <td>&gt;1 nonsynonymous</td>
      <td>-12.580142</td>
    </tr>
    <tr>
      <th>181366</th>
      <td>lib_2</td>
      <td>h2</td>
      <td>F15R I35S D38E Y39G</td>
      <td>tight_bottle</td>
      <td>-5.781253</td>
      <td>&gt;1 nonsynonymous</td>
      <td>-15.097683</td>
    </tr>
    <tr>
      <th>181367</th>
      <td>lib_2</td>
      <td>h2</td>
      <td>E9F C34P Y39C E44Y T45F</td>
      <td>tight_bottle</td>
      <td>-5.581944</td>
      <td>&gt;1 nonsynonymous</td>
      <td>-6.628840</td>
    </tr>
  </tbody>
</table>
<p>181368 rows × 7 columns</p>
</div></div></div>
</div>
<p>Let’s plot the functional scores as a function of the latent phenotype.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">p</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">ggplot</span><span class="p">(</span><span class="n">func_scores</span><span class="p">,</span> <span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="s2">&quot;latent_phenotype&quot;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s2">&quot;func_score&quot;</span><span class="p">))</span>
    <span class="o">+</span> <span class="n">geom_point</span><span class="p">(</span><span class="n">alpha</span><span class="o">=</span><span class="mf">0.05</span><span class="p">)</span>
    <span class="o">+</span> <span class="n">facet_wrap</span><span class="p">(</span><span class="s2">&quot;~func_score_type&quot;</span><span class="p">)</span>
    <span class="o">+</span> <span class="n">theme_classic</span><span class="p">()</span>
    <span class="o">+</span> <span class="n">theme</span><span class="p">(</span><span class="n">figure_size</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="mi">3</span><span class="p">))</span>
<span class="p">)</span>
<span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">(</span><span class="n">show</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/10db1e7399c4c26f1f3a227a4d72bb0f613b1bcbe6ccde1c10c4cd3185f6c407.png" src="_images/10db1e7399c4c26f1f3a227a4d72bb0f613b1bcbe6ccde1c10c4cd3185f6c407.png" />
</div>
</div>
<p>Plot a pairplot to see how targets compare.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">pairplot</span><span class="p">(</span>
    <span class="p">(</span>
        <span class="n">func_scores</span>
        <span class="o">.</span><span class="n">pivot</span><span class="p">(</span>
            <span class="n">index</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;library&quot;</span><span class="p">,</span> <span class="s2">&quot;homolog&quot;</span><span class="p">,</span> <span class="s2">&quot;aa_substitutions&quot;</span><span class="p">,</span> <span class="s2">&quot;variant_class&quot;</span><span class="p">],</span>
            <span class="n">columns</span><span class="o">=</span><span class="s2">&quot;func_score_type&quot;</span><span class="p">,</span>
            <span class="n">values</span><span class="o">=</span><span class="s2">&quot;func_score&quot;</span>
        <span class="p">)</span>
        <span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
        <span class="o">.</span><span class="n">sample</span><span class="p">(</span><span class="n">frac</span><span class="o">=</span><span class="mf">0.1</span><span class="p">)</span>
    <span class="p">),</span>  
    <span class="n">hue</span><span class="o">=</span><span class="s1">&#39;homolog&#39;</span><span class="p">,</span>
    <span class="n">plot_kws</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;alpha&quot;</span><span class="p">:</span><span class="mf">0.25</span><span class="p">},</span>
    <span class="n">corner</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>
<span class="n">fig</span><span class="o">.</span><span class="n">fig</span><span class="o">.</span><span class="n">set_size_inches</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="mi">5</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/0f8fd953d80ccb7b9b3ce08ad1dbd87981d20f52222a0920874f75c8a843453a.png" src="_images/0f8fd953d80ccb7b9b3ce08ad1dbd87981d20f52222a0920874f75c8a843453a.png" />
</div>
</div>
<p>We can see here that the tight bottleneck indeed introduces more noise in the data, as the functional scores are less correlated with the ground truth phenotype.</p>
<p>Plot the functional scores distributions of the bottleneck counts-computed functional scores. Recall that we collapsed barcode replicates so there is only a single wildtype phenotype in each of the datasets, thus we will not plot them below.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">p</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">ggplot</span><span class="p">(</span>
        <span class="n">func_scores</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s2">&quot;(func_score_type != &#39;observed_phenotype&#39;) &amp; (variant_class != &#39;wildtype&#39;)&quot;</span><span class="p">),</span> 
        <span class="n">aes</span><span class="p">(</span><span class="s2">&quot;variant_class&quot;</span><span class="p">,</span> <span class="s2">&quot;func_score&quot;</span><span class="p">)</span>
    <span class="p">)</span>
    <span class="o">+</span> <span class="n">geom_violin</span><span class="p">(</span><span class="n">fill</span><span class="o">=</span><span class="s2">&quot;lightblue&quot;</span><span class="p">)</span>
    <span class="o">+</span> <span class="n">ylab</span><span class="p">(</span><span class="s2">&quot;functional score&quot;</span><span class="p">)</span>
    <span class="o">+</span> <span class="n">xlab</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
    <span class="o">+</span> <span class="n">facet_grid</span><span class="p">(</span><span class="s2">&quot;func_score_type ~ library + homolog&quot;</span><span class="p">)</span>
    <span class="o">+</span> <span class="n">theme_classic</span><span class="p">()</span>
    <span class="o">+</span> <span class="n">theme</span><span class="p">(</span>
        <span class="n">figure_size</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="mi">5</span><span class="p">),</span>
        <span class="n">axis_text_x</span><span class="o">=</span><span class="n">element_text</span><span class="p">(</span><span class="n">angle</span><span class="o">=</span><span class="mi">90</span><span class="p">),</span>
        <span class="n">panel_grid_major_x</span><span class="o">=</span><span class="n">element_blank</span><span class="p">(),</span>  <span class="c1"># no vertical grid lines</span>
    <span class="p">)</span>
    <span class="c1"># + scale_fill_manual(values=CBPALETTE[1:])</span>
<span class="p">)</span>
<span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">(</span><span class="n">show</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/46d47a844a5a1cd88ebdbdfd4e7d8def298827edbb282a4530440e7d22a04789.png" src="_images/46d47a844a5a1cd88ebdbdfd4e7d8def298827edbb282a4530440e7d22a04789.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">func_scores</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">csv_output_dir</span><span class="si">}</span><span class="s2">/simulated_func_scores.csv&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">func_scores</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>library</th>
      <th>homolog</th>
      <th>aa_substitutions</th>
      <th>func_score_type</th>
      <th>func_score</th>
      <th>variant_class</th>
      <th>latent_phenotype</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>lib_1</td>
      <td>h1</td>
      <td>V24Q S25R H27P D29L H33C S41L C44F S47F</td>
      <td>observed_phenotype</td>
      <td>-5.96</td>
      <td>&gt;1 nonsynonymous</td>
      <td>-11.10</td>
    </tr>
    <tr>
      <th>1</th>
      <td>lib_1</td>
      <td>h1</td>
      <td>V8P H27T P30* L49P</td>
      <td>observed_phenotype</td>
      <td>-5.96</td>
      <td>stop</td>
      <td>-11.11</td>
    </tr>
    <tr>
      <th>2</th>
      <td>lib_1</td>
      <td>h1</td>
      <td>V19P V43Q</td>
      <td>observed_phenotype</td>
      <td>-5.64</td>
      <td>&gt;1 nonsynonymous</td>
      <td>-2.88</td>
    </tr>
    <tr>
      <th>3</th>
      <td>lib_1</td>
      <td>h1</td>
      <td>R20Y</td>
      <td>observed_phenotype</td>
      <td>0.01</td>
      <td>1 nonsynonymous</td>
      <td>5.16</td>
    </tr>
    <tr>
      <th>4</th>
      <td>lib_1</td>
      <td>h1</td>
      <td></td>
      <td>observed_phenotype</td>
      <td>0.00</td>
      <td>wildtype</td>
      <td>5.00</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
</section>
<section id="create-data-objects-for-each-library-replicate-and-bottleneck">
<h2>Create <code class="docutils literal notranslate"><span class="pre">Data</span></code> objects for each library replicate, and bottleneck<a class="headerlink" href="#create-data-objects-for-each-library-replicate-and-bottleneck" title="Link to this heading">¶</a></h2>
<p><code class="docutils literal notranslate"><span class="pre">multidms</span></code> model fitting is generally a two-step process: (1) Create <code class="docutils literal notranslate"><span class="pre">Data</span></code> objects from functional score dataframes which, among other things, encode the variant data into one-hot encoded matrices, and (2) fit a collection of models across a grid of <code class="docutils literal notranslate"><span class="pre">Data</span></code> objects and hyperparameters. Here, we’ll create <code class="docutils literal notranslate"><span class="pre">Data</span></code> objects for each library replicate / fitting target combination:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">data_objects</span> <span class="o">=</span> <span class="p">[</span>
    <span class="n">multidms</span><span class="o">.</span><span class="n">Data</span><span class="p">(</span>
        <span class="n">fs_df</span><span class="p">,</span> 
        <span class="n">reference</span><span class="o">=</span><span class="s2">&quot;h1&quot;</span><span class="p">,</span> 
        <span class="n">alphabet</span> <span class="o">=</span> <span class="n">multidms</span><span class="o">.</span><span class="n">AAS_WITHSTOP_WITHGAP</span><span class="p">,</span> 
        <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">name</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">lib</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">target</span><span class="si">}</span><span class="s2">_func_score&quot;</span>
    <span class="p">)</span>
    <span class="k">for</span> <span class="p">(</span><span class="n">lib</span><span class="p">,</span> <span class="n">target</span><span class="p">),</span> <span class="n">fs_df</span> <span class="ow">in</span> <span class="n">func_scores</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;homolog&quot;</span><span class="p">:</span><span class="s2">&quot;condition&quot;</span><span class="p">})</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;library&#39;</span><span class="p">,</span> <span class="s1">&#39;func_score_type&#39;</span><span class="p">])</span>
<span class="p">]</span>

<span class="n">data_objects</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[Data(lib_1_observed_phenotype_func_score),
 Data(lib_1_loose_bottle_func_score),
 Data(lib_1_tight_bottle_func_score),
 Data(lib_2_observed_phenotype_func_score),
 Data(lib_2_loose_bottle_func_score),
 Data(lib_2_tight_bottle_func_score)]
</pre></div>
</div>
</div>
</div>
</section>
<section id="fit-models-to-training-data-lasso-sweep">
<h2>Fit models to training data (lasso sweep)<a class="headerlink" href="#fit-models-to-training-data-lasso-sweep" title="Link to this heading">¶</a></h2>
<p>Next, we’ll fit a set of models to each of the datasets defined above using the <code class="docutils literal notranslate"><span class="pre">multidms.fit_models</span></code> <a class="reference external" href="https://matsengrp.github.io/multidms/multidms.model_collection.html#multidms.model_collection.fit_models">function</a>. For each dataset, we independently fit models with a number of different lasso penalty coefficients ($\lambda_{L1}$) as defined at the top of this notebook.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fitting_params</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;num_training_steps&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">num_training_steps</span><span class="p">],</span> <span class="c1"># default 100</span>
    <span class="s2">&quot;iterations_per_step&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">iterations_per_step</span><span class="p">],</span> <span class="c1"># default 20000</span>
    <span class="s2">&quot;scale_coeff_lasso_shift&quot;</span><span class="p">:</span> <span class="n">scale_coeff_lasso_shift</span><span class="p">,</span> <span class="c1"># the sweep of lasso coefficient params</span>
    <span class="s2">&quot;init_beta_naught&quot;</span> <span class="p">:</span> <span class="p">[</span><span class="n">init_beta_naught</span><span class="p">],</span> <span class="c1"># We&#39;ve found that we need to start with a higher beta_naught to get the model to converge correctly,</span>
    <span class="s2">&quot;scale_coeff_ridge_beta&quot;</span> <span class="p">:</span> <span class="p">[</span><span class="n">scale_coeff_ridge_beta</span><span class="p">],</span> <span class="c1"># the sweep of ridge coefficient params</span>
<span class="p">}</span>

<span class="n">fitting_params</span><span class="p">[</span><span class="s2">&quot;dataset&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">data_objects</span>
<span class="n">pprint</span><span class="o">.</span><span class="n">pprint</span><span class="p">(</span><span class="n">fitting_params</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>{&#39;dataset&#39;: [Data(lib_1_observed_phenotype_func_score),
             Data(lib_1_loose_bottle_func_score),
             Data(lib_1_tight_bottle_func_score),
             Data(lib_2_observed_phenotype_func_score),
             Data(lib_2_loose_bottle_func_score),
             Data(lib_2_tight_bottle_func_score)],
 &#39;init_beta_naught&#39;: [5.0],
 &#39;iterations_per_step&#39;: [100000],
 &#39;num_training_steps&#39;: [1],
 &#39;scale_coeff_lasso_shift&#39;: [0.0,
                             5e-06,
                             1e-05,
                             2e-05,
                             4e-05,
                             8e-05,
                             0.00016,
                             0.00032,
                             0.00064],
 &#39;scale_coeff_ridge_beta&#39;: [1e-07]}
</pre></div>
</div>
</div>
</div>
<p>Fit the models:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">_</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">fit_collection_df</span> <span class="o">=</span> <span class="n">multidms</span><span class="o">.</span><span class="n">model_collection</span><span class="o">.</span><span class="n">fit_models</span><span class="p">(</span><span class="n">fitting_params</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Note that the return type of <a class="reference external" href="https://matsengrp.github.io/multidms/multidms.html#multidms.fit_models">multidms.fit_models</a> is a <code class="docutils literal notranslate"><span class="pre">tuple(int,</span> <span class="pre">int,</span> <span class="pre">pd.DataFrame)</span></code> where the first two values are the number of fits that were successful and the number that failed, respectively, and the third value is a dataframe where each row contains a fit <code class="docutils literal notranslate"><span class="pre">Model</span></code> object and the hyperparameters used to fit it.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fit_collection_df</span><span class="p">[[</span><span class="s2">&quot;model&quot;</span><span class="p">,</span> <span class="s2">&quot;dataset_name&quot;</span><span class="p">,</span> <span class="s2">&quot;scale_coeff_lasso_shift&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>model</th>
      <th>dataset_name</th>
      <th>scale_coeff_lasso_shift</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>Model(Model-0)</td>
      <td>lib_1_observed_phenotype_func_score</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>1</th>
      <td>Model(Model-0)</td>
      <td>lib_1_observed_phenotype_func_score</td>
      <td>0.000005</td>
    </tr>
    <tr>
      <th>2</th>
      <td>Model(Model-0)</td>
      <td>lib_1_observed_phenotype_func_score</td>
      <td>0.00001</td>
    </tr>
    <tr>
      <th>3</th>
      <td>Model(Model-0)</td>
      <td>lib_1_observed_phenotype_func_score</td>
      <td>0.00002</td>
    </tr>
    <tr>
      <th>4</th>
      <td>Model(Model-0)</td>
      <td>lib_1_observed_phenotype_func_score</td>
      <td>0.00004</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>Add a few helpful features to this dataframe for plotting by splitting the “dataset_name” (name of the Data Object that was used for fitting) into more understandable columns:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fit_collection_df</span> <span class="o">=</span> <span class="n">fit_collection_df</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span>
    <span class="n">measurement_type</span> <span class="o">=</span> <span class="n">fit_collection_df</span><span class="p">[</span><span class="s2">&quot;dataset_name&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">str</span><span class="p">[</span><span class="mi">2</span><span class="p">:</span><span class="mi">4</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">),</span>
    <span class="n">library</span> <span class="o">=</span> <span class="n">fit_collection_df</span><span class="p">[</span><span class="s2">&quot;dataset_name&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">str</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)</span>
<span class="p">)</span>

<span class="c1"># convert measurement type to categorical ordered by &#39;observed&#39;, &#39;loose&#39;, &#39;tight&#39; for plotting</span>
<span class="n">fit_collection_df</span><span class="p">[</span><span class="s2">&quot;measurement_type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Categorical</span><span class="p">(</span>
    <span class="n">fit_collection_df</span><span class="p">[</span><span class="s2">&quot;measurement_type&quot;</span><span class="p">],</span>
    <span class="n">categories</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;observed_phenotype&quot;</span><span class="p">,</span> <span class="s2">&quot;loose_bottle&quot;</span><span class="p">,</span> <span class="s2">&quot;tight_bottle&quot;</span><span class="p">],</span>
    <span class="n">ordered</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>

<span class="n">fit_collection_df</span><span class="p">[[</span><span class="s2">&quot;model&quot;</span><span class="p">,</span> <span class="s2">&quot;measurement_type&quot;</span><span class="p">,</span> <span class="s2">&quot;library&quot;</span><span class="p">,</span> <span class="s2">&quot;scale_coeff_lasso_shift&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>model</th>
      <th>measurement_type</th>
      <th>library</th>
      <th>scale_coeff_lasso_shift</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>Model(Model-0)</td>
      <td>observed_phenotype</td>
      <td>lib_1</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>1</th>
      <td>Model(Model-0)</td>
      <td>observed_phenotype</td>
      <td>lib_1</td>
      <td>0.000005</td>
    </tr>
    <tr>
      <th>2</th>
      <td>Model(Model-0)</td>
      <td>observed_phenotype</td>
      <td>lib_1</td>
      <td>0.00001</td>
    </tr>
    <tr>
      <th>3</th>
      <td>Model(Model-0)</td>
      <td>observed_phenotype</td>
      <td>lib_1</td>
      <td>0.00002</td>
    </tr>
    <tr>
      <th>4</th>
      <td>Model(Model-0)</td>
      <td>observed_phenotype</td>
      <td>lib_1</td>
      <td>0.00004</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
</section>
<section id="model-vs-truth-mutational-effects">
<h2>Model vs. truth mutational effects<a class="headerlink" href="#model-vs-truth-mutational-effects" title="Link to this heading">¶</a></h2>
<p>The <code class="docutils literal notranslate"><span class="pre">multidms.ModelCollection</span></code> takes a a dataframe such as the one created above and provides a few helpful methods for model selection and aggregation. We’ll use these methods to evaluate the fits and select the best model for each dataset.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">model_collection</span> <span class="o">=</span> <span class="n">multidms</span><span class="o">.</span><span class="n">model_collection</span><span class="o">.</span><span class="n">ModelCollection</span><span class="p">(</span><span class="n">fit_collection_df</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Get the mutational parameters for each of the models using <code class="docutils literal notranslate"><span class="pre">model_collection.split_apply_combine_muts()</span></code> <a class="reference external" href="https://matsengrp.github.io/multidms/multidms.model_collection.html#multidms.model_collection.ModelCollection.split_apply_combine_muts">method</a>, and merge them with the simulated ground truth parameters:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># the columns that distinguish fits are what we&#39;ll groupby such that none of the model collection parameters are aggregated</span>
<span class="n">groupby</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;library&quot;</span><span class="p">,</span> <span class="s2">&quot;measurement_type&quot;</span><span class="p">,</span> <span class="s2">&quot;scale_coeff_lasso_shift&quot;</span><span class="p">)</span>
<span class="n">collection_muts_df</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">model_collection</span><span class="o">.</span><span class="n">split_apply_combine_muts</span><span class="p">(</span>
        <span class="n">groupby</span><span class="o">=</span><span class="n">groupby</span>
    <span class="p">)</span>
    <span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
    <span class="o">.</span><span class="n">rename</span><span class="p">(</span>
        <span class="p">{</span>
            <span class="s1">&#39;beta&#39;</span> <span class="p">:</span> <span class="s1">&#39;predicted_beta&#39;</span><span class="p">,</span>
            <span class="s1">&#39;shift_h2&#39;</span> <span class="p">:</span> <span class="s1">&#39;predicted_shift_h2&#39;</span><span class="p">,</span>
        <span class="p">},</span> 
        <span class="n">axis</span><span class="o">=</span><span class="mi">1</span>
    <span class="p">)</span>    
    <span class="o">.</span><span class="n">merge</span><span class="p">(</span>
        <span class="n">mut_effects_df</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span>
            <span class="p">{</span>
                <span class="s1">&#39;beta_h1&#39;</span> <span class="p">:</span> <span class="s1">&#39;true_beta&#39;</span><span class="p">,</span>
                <span class="s1">&#39;beta_h2&#39;</span> <span class="p">:</span> <span class="s1">&#39;true_beta_h2&#39;</span><span class="p">,</span>
                <span class="s1">&#39;shift&#39;</span> <span class="p">:</span> <span class="s1">&#39;true_shift&#39;</span><span class="p">,</span>
            <span class="p">},</span>
            <span class="n">axis</span><span class="o">=</span><span class="mi">1</span>
        <span class="p">),</span>
        <span class="n">on</span><span class="o">=</span><span class="s1">&#39;mutation&#39;</span>
    <span class="p">)</span>
<span class="p">)</span>
<span class="k">assert</span> <span class="n">collection_muts_df</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">mut_effects_df</span><span class="p">)</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">fit_collection_df</span><span class="p">)</span>
<span class="n">collection_muts_df</span><span class="p">[[</span><span class="s2">&quot;mutation&quot;</span><span class="p">,</span> <span class="s2">&quot;library&quot;</span><span class="p">,</span> <span class="s2">&quot;measurement_type&quot;</span><span class="p">,</span> <span class="s2">&quot;scale_coeff_lasso_shift&quot;</span><span class="p">,</span> <span class="s2">&quot;predicted_shift_h2&quot;</span><span class="p">,</span> <span class="s2">&quot;true_shift&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>cache miss - this could take a moment
</pre></div>
</div>
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>mutation</th>
      <th>library</th>
      <th>measurement_type</th>
      <th>scale_coeff_lasso_shift</th>
      <th>predicted_shift_h2</th>
      <th>true_shift</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>C31*</td>
      <td>lib_1</td>
      <td>observed_phenotype</td>
      <td>0.0</td>
      <td>-15.289572</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>1</th>
      <td>C31A</td>
      <td>lib_1</td>
      <td>observed_phenotype</td>
      <td>0.0</td>
      <td>-0.003453</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>2</th>
      <td>C31D</td>
      <td>lib_1</td>
      <td>observed_phenotype</td>
      <td>0.0</td>
      <td>-0.030438</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>3</th>
      <td>C31E</td>
      <td>lib_1</td>
      <td>observed_phenotype</td>
      <td>0.0</td>
      <td>-0.086455</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>4</th>
      <td>C31F</td>
      <td>lib_1</td>
      <td>observed_phenotype</td>
      <td>0.0</td>
      <td>-0.021350</td>
      <td>0.0</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>To compare model fits parameter values to the simulated ground truth values, add mean squared error and pearsonr metrics to the <code class="docutils literal notranslate"><span class="pre">ModelCollection.fit_models</span></code> attribute:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">series_corr</span><span class="p">(</span><span class="n">y_true</span><span class="p">,</span> <span class="n">y_pred</span><span class="p">):</span>
    <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">corrcoef</span><span class="p">(</span><span class="n">y_true</span><span class="p">,</span> <span class="n">y_pred</span><span class="p">)[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span>

<span class="k">def</span> <span class="nf">series_mae</span><span class="p">(</span><span class="n">y_true</span><span class="p">,</span> <span class="n">y_pred</span><span class="p">):</span>
    <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">y_true</span> <span class="o">-</span> <span class="n">y_pred</span><span class="p">)</span>

<span class="c1"># compute the new metric columns</span>
<span class="n">new_fit_models_cols</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">list</span><span class="p">)</span>
<span class="k">for</span> <span class="n">group</span><span class="p">,</span> <span class="n">model_mutations_df</span> <span class="ow">in</span> <span class="n">collection_muts_df</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">groupby</span><span class="p">)):</span>
    <span class="c1"># add cols for merging</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">attribute</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">group</span><span class="p">):</span>
        <span class="n">new_fit_models_cols</span><span class="p">[</span><span class="n">groupby</span><span class="p">[</span><span class="n">i</span><span class="p">]]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">group</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
        
    <span class="k">for</span> <span class="n">parameter</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;beta&quot;</span><span class="p">,</span> <span class="s2">&quot;shift&quot;</span><span class="p">]:</span>
        <span class="k">for</span> <span class="n">metric_fxn</span><span class="p">,</span> <span class="n">name</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">([</span><span class="n">series_corr</span><span class="p">,</span> <span class="n">series_mae</span><span class="p">],</span> <span class="p">[</span><span class="s2">&quot;corr&quot;</span><span class="p">,</span> <span class="s2">&quot;mae&quot;</span><span class="p">]):</span>

            <span class="c1"># add the new metric columns</span>
            <span class="n">postfix</span><span class="o">=</span><span class="s2">&quot;_h2&quot;</span> <span class="k">if</span> <span class="n">parameter</span> <span class="o">==</span> <span class="s2">&quot;shift&quot;</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span>
            <span class="n">y_pred</span> <span class="o">=</span> <span class="n">model_mutations_df</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;predicted_</span><span class="si">{</span><span class="n">parameter</span><span class="si">}{</span><span class="n">postfix</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]</span>
            <span class="n">y_true</span> <span class="o">=</span> <span class="n">model_mutations_df</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;true_</span><span class="si">{</span><span class="n">parameter</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]</span>
            <span class="n">new_fit_models_cols</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">parameter</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">metric_fxn</span><span class="p">(</span><span class="n">y_true</span><span class="p">,</span> <span class="n">y_pred</span><span class="p">))</span>

<span class="c1"># merge the new columns into the model collection</span>
<span class="n">model_collection</span><span class="o">.</span><span class="n">fit_models</span> <span class="o">=</span> <span class="n">model_collection</span><span class="o">.</span><span class="n">fit_models</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span>
    <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">new_fit_models_cols</span><span class="p">),</span>
    <span class="n">on</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="n">groupby</span><span class="p">)</span>
<span class="p">)</span>

<span class="c1"># print the first few rows of the model collection</span>
<span class="n">model_collection</span><span class="o">.</span><span class="n">fit_models</span><span class="p">[</span><span class="nb">list</span><span class="p">(</span><span class="n">groupby</span><span class="p">)</span> <span class="o">+</span> <span class="p">[</span><span class="n">c</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">new_fit_models_cols</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span> <span class="k">if</span> <span class="n">c</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">groupby</span><span class="p">]]</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>library</th>
      <th>measurement_type</th>
      <th>scale_coeff_lasso_shift</th>
      <th>beta_corr</th>
      <th>beta_mae</th>
      <th>shift_corr</th>
      <th>shift_mae</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>lib_1</td>
      <td>observed_phenotype</td>
      <td>0.0</td>
      <td>0.977246</td>
      <td>-0.268842</td>
      <td>0.109992</td>
      <td>1.172704</td>
    </tr>
    <tr>
      <th>1</th>
      <td>lib_1</td>
      <td>observed_phenotype</td>
      <td>0.000005</td>
      <td>0.985431</td>
      <td>-0.206331</td>
      <td>0.997367</td>
      <td>0.000190</td>
    </tr>
    <tr>
      <th>2</th>
      <td>lib_1</td>
      <td>observed_phenotype</td>
      <td>0.00001</td>
      <td>0.984181</td>
      <td>-0.217609</td>
      <td>0.994465</td>
      <td>-0.000601</td>
    </tr>
    <tr>
      <th>3</th>
      <td>lib_1</td>
      <td>observed_phenotype</td>
      <td>0.00002</td>
      <td>0.984685</td>
      <td>-0.220858</td>
      <td>0.990086</td>
      <td>-0.000248</td>
    </tr>
    <tr>
      <th>4</th>
      <td>lib_1</td>
      <td>observed_phenotype</td>
      <td>0.00004</td>
      <td>0.984756</td>
      <td>-0.226117</td>
      <td>0.979619</td>
      <td>0.001351</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>Next, we’ll make some summary plots with the predicted vs. simulated ground MSE truth parameters across model fits</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">metric</span> <span class="o">=</span> <span class="s2">&quot;corr&quot;</span>

<span class="n">data</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">model_collection</span><span class="o">.</span><span class="n">fit_models</span>
    <span class="o">.</span><span class="n">assign</span><span class="p">(</span>
        <span class="n">measurement_library</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;measurement_type&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot; &quot;</span> <span class="o">+</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;library&quot;</span><span class="p">]</span>
    <span class="p">)</span>
    <span class="o">.</span><span class="n">melt</span><span class="p">(</span>
        <span class="n">id_vars</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="n">groupby</span><span class="p">)</span> <span class="o">+</span> <span class="p">[</span><span class="s2">&quot;measurement_library&quot;</span><span class="p">],</span>
        <span class="n">value_vars</span><span class="o">=</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;beta_</span><span class="si">{</span><span class="n">metric</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;shift_</span><span class="si">{</span><span class="n">metric</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">],</span>
        <span class="n">var_name</span><span class="o">=</span><span class="s2">&quot;parameter&quot;</span><span class="p">,</span>
        <span class="n">value_name</span><span class="o">=</span><span class="n">metric</span><span class="p">,</span>
    <span class="p">)</span>
<span class="p">)</span>
<span class="n">data</span><span class="p">[</span><span class="s2">&quot;measurement_type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Categorical</span><span class="p">(</span>
    <span class="n">data</span><span class="p">[</span><span class="s2">&quot;measurement_type&quot;</span><span class="p">],</span>
    <span class="n">categories</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;observed_phenotype&quot;</span><span class="p">,</span> <span class="s2">&quot;loose_bottle&quot;</span><span class="p">,</span> <span class="s2">&quot;tight_bottle&quot;</span><span class="p">],</span>
    <span class="n">ordered</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>
<span class="n">data</span><span class="p">[</span><span class="s2">&quot;parameter&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;parameter&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;_</span><span class="si">{</span><span class="n">metric</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="n">data</span><span class="p">[</span><span class="s2">&quot;parameter&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Categorical</span><span class="p">(</span>
    <span class="n">data</span><span class="p">[</span><span class="s2">&quot;parameter&quot;</span><span class="p">],</span>
    <span class="n">categories</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;shift&quot;</span><span class="p">,</span> <span class="s2">&quot;beta&quot;</span><span class="p">],</span>
    <span class="n">ordered</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>
<span class="k">for</span> <span class="n">parameter</span><span class="p">,</span> <span class="n">parameter_df</span> <span class="ow">in</span> <span class="n">data</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;parameter&quot;</span><span class="p">,</span> <span class="n">observed</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
    <span class="n">p</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">ggplot</span><span class="p">(</span><span class="n">parameter_df</span><span class="p">)</span>
        <span class="o">+</span> <span class="n">geom_line</span><span class="p">(</span>
            <span class="n">aes</span><span class="p">(</span>
                <span class="n">x</span><span class="o">=</span><span class="s2">&quot;scale_coeff_lasso_shift&quot;</span><span class="p">,</span>
                <span class="n">y</span><span class="o">=</span><span class="n">metric</span><span class="p">,</span>
                <span class="n">group</span><span class="o">=</span><span class="s2">&quot;measurement_library&quot;</span><span class="p">,</span>
            <span class="p">),</span>
        <span class="p">)</span>
        <span class="o">+</span> <span class="n">geom_point</span><span class="p">(</span>
            <span class="n">aes</span><span class="p">(</span>
                <span class="n">x</span><span class="o">=</span><span class="s2">&quot;scale_coeff_lasso_shift&quot;</span><span class="p">,</span>
                <span class="n">y</span><span class="o">=</span><span class="n">metric</span><span class="p">,</span>
                <span class="n">shape</span><span class="o">=</span><span class="s2">&quot;library&quot;</span>
            <span class="p">),</span>
            <span class="n">size</span><span class="o">=</span><span class="mi">4</span>
        <span class="p">)</span>
        <span class="o">+</span> <span class="n">facet_wrap</span><span class="p">(</span><span class="s2">&quot;measurement_type&quot;</span><span class="p">,</span> <span class="n">scales</span><span class="o">=</span><span class="s2">&quot;free_y&quot;</span><span class="p">)</span>
        <span class="o">+</span> <span class="n">theme_classic</span><span class="p">()</span>
        <span class="o">+</span> <span class="n">theme</span><span class="p">(</span>
            <span class="n">figure_size</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="mf">3.3</span><span class="p">),</span>
            <span class="n">axis_text_x</span><span class="o">=</span><span class="n">element_text</span><span class="p">(</span><span class="n">angle</span><span class="o">=</span><span class="mi">90</span><span class="p">),</span>
            <span class="n">panel_grid_major_x</span><span class="o">=</span><span class="n">element_blank</span><span class="p">(),</span>  <span class="c1"># no vertical grid lines</span>
        <span class="p">)</span>
        <span class="o">+</span> <span class="n">labs</span><span class="p">(</span>
            <span class="n">title</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;Prediction vs. Simulated Ground Truth: </span><span class="si">{</span><span class="n">parameter</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>
            <span class="n">x</span><span class="o">=</span><span class="s2">&quot;lasso penalty coefficient (λ)&quot;</span><span class="p">,</span>
            <span class="n">y</span><span class="o">=</span><span class="s2">&quot;Mean Absolute Error&quot;</span> <span class="k">if</span> <span class="n">metric</span> <span class="o">==</span> <span class="s2">&quot;mae&quot;</span> <span class="k">else</span> <span class="s2">&quot;Correlation&quot;</span>
        <span class="p">)</span>
    <span class="p">)</span>
    <span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">(</span><span class="n">show</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/c91326a0b46bed2d2b706e61ec20059a3bcc7fcc936a5cc9b1c497c1648954e0.png" src="_images/c91326a0b46bed2d2b706e61ec20059a3bcc7fcc936a5cc9b1c497c1648954e0.png" />
<img alt="_images/5f33b25f036aa8f55833f6269e5c9709d4c2d23dd1cff5343a707afbaa276e84.png" src="_images/5f33b25f036aa8f55833f6269e5c9709d4c2d23dd1cff5343a707afbaa276e84.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">data</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">csv_output_dir</span><span class="si">}</span><span class="s2">/model_vs_truth_beta_shift.csv&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">data</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>library</th>
      <th>measurement_type</th>
      <th>scale_coeff_lasso_shift</th>
      <th>measurement_library</th>
      <th>parameter</th>
      <th>corr</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>lib_1</td>
      <td>observed_phenotype</td>
      <td>0.0</td>
      <td>observed_phenotype lib_1</td>
      <td>beta</td>
      <td>0.98</td>
    </tr>
    <tr>
      <th>1</th>
      <td>lib_1</td>
      <td>observed_phenotype</td>
      <td>0.000005</td>
      <td>observed_phenotype lib_1</td>
      <td>beta</td>
      <td>0.99</td>
    </tr>
    <tr>
      <th>2</th>
      <td>lib_1</td>
      <td>observed_phenotype</td>
      <td>0.00001</td>
      <td>observed_phenotype lib_1</td>
      <td>beta</td>
      <td>0.98</td>
    </tr>
    <tr>
      <th>3</th>
      <td>lib_1</td>
      <td>observed_phenotype</td>
      <td>0.00002</td>
      <td>observed_phenotype lib_1</td>
      <td>beta</td>
      <td>0.98</td>
    </tr>
    <tr>
      <th>4</th>
      <td>lib_1</td>
      <td>observed_phenotype</td>
      <td>0.00004</td>
      <td>observed_phenotype lib_1</td>
      <td>beta</td>
      <td>0.98</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
</section>
<section id="shift-sparsity">
<h2>Shift sparsity<a class="headerlink" href="#shift-sparsity" title="Link to this heading">¶</a></h2>
<p>Another way we might evaluate the fits is by looking at the <em>sparsity</em> of the models by computing the percentage of shift parameters that are equal to zero among the models. We look at this metric separately for mutations to stop codons and mutations to non-stop codons because we expect the stop codons to be equally deleterious in both homologs, and thus we expect all the shift parameters associated with stop codon mutations to be zero in a “good” fit.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">chart</span><span class="p">,</span> <span class="n">data</span> <span class="o">=</span> <span class="n">model_collection</span><span class="o">.</span><span class="n">shift_sparsity</span><span class="p">(</span><span class="n">return_data</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">data</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>cache miss - this could take a moment
</pre></div>
</div>
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>dataset_name</th>
      <th>scale_coeff_lasso_shift</th>
      <th>mut_type</th>
      <th>mut_param</th>
      <th>sparsity</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>lib_1_loose_bottle_func_score</td>
      <td>0.000000</td>
      <td>nonsynonymous</td>
      <td>shift_h2</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>1</th>
      <td>lib_1_loose_bottle_func_score</td>
      <td>0.000000</td>
      <td>stop</td>
      <td>shift_h2</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>2</th>
      <td>lib_1_loose_bottle_func_score</td>
      <td>0.000005</td>
      <td>nonsynonymous</td>
      <td>shift_h2</td>
      <td>0.078947</td>
    </tr>
    <tr>
      <th>3</th>
      <td>lib_1_loose_bottle_func_score</td>
      <td>0.000005</td>
      <td>stop</td>
      <td>shift_h2</td>
      <td>0.660000</td>
    </tr>
    <tr>
      <th>4</th>
      <td>lib_1_loose_bottle_func_score</td>
      <td>0.000010</td>
      <td>nonsynonymous</td>
      <td>shift_h2</td>
      <td>0.148421</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">data</span> <span class="o">=</span><span class="n">data</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span>
    <span class="n">library</span><span class="o">=</span><span class="n">data</span><span class="o">.</span><span class="n">dataset_name</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">str</span><span class="p">[:</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">),</span>
    <span class="n">library_type</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">dataset_name</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">str</span><span class="p">[:</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;-&quot;</span> <span class="o">+</span> <span class="n">data</span><span class="o">.</span><span class="n">mut_type</span><span class="p">,</span>
    <span class="n">measurement_type</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">dataset_name</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">str</span><span class="p">[</span><span class="mi">2</span><span class="p">:</span><span class="mi">4</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)</span>
<span class="p">)</span>
<span class="c1"># convert measurement type to categorical ordered by &#39;observed&#39;, &#39;loose&#39;, &#39;tight&#39;</span>
<span class="n">data</span><span class="p">[</span><span class="s2">&quot;measurement_type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Categorical</span><span class="p">(</span>
    <span class="n">data</span><span class="p">[</span><span class="s2">&quot;measurement_type&quot;</span><span class="p">],</span>
    <span class="n">categories</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;observed_phenotype&quot;</span><span class="p">,</span> <span class="s2">&quot;loose_bottle&quot;</span><span class="p">,</span> <span class="s2">&quot;tight_bottle&quot;</span><span class="p">],</span>
    <span class="n">ordered</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>
<span class="n">data</span><span class="p">[</span><span class="s1">&#39;True Sparsity&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span> <span class="c1">#dummy col for plotting legend</span>
<span class="c1"># data.sort_values(&quot;scale_coeff_lasso_shift&quot;, inplace=True)</span>
<span class="n">data</span><span class="p">[</span><span class="s2">&quot;scale_coeff_lasso_shift&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">scale_coeff_lasso_shift</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">object</span><span class="p">)</span>

<span class="n">true_sparsity</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">-</span> <span class="n">mut_effects_df</span><span class="o">.</span><span class="n">shifted_site</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>

<span class="n">p</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">ggplot</span><span class="p">(</span>
        <span class="n">data</span><span class="p">,</span>
        <span class="n">aes</span><span class="p">(</span><span class="s2">&quot;scale_coeff_lasso_shift&quot;</span><span class="p">,</span> <span class="s2">&quot;sparsity&quot;</span><span class="p">)</span>
    <span class="p">)</span>
    <span class="o">+</span> <span class="n">geom_hline</span><span class="p">(</span><span class="n">aes</span><span class="p">(</span><span class="n">yintercept</span><span class="o">=</span><span class="n">true_sparsity</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;True Sparsity&quot;</span><span class="p">),</span> <span class="n">linetype</span><span class="o">=</span><span class="s2">&quot;dashed&quot;</span><span class="p">)</span>
    <span class="o">+</span> <span class="n">geom_line</span><span class="p">(</span>
        <span class="n">aes</span><span class="p">(</span>
            <span class="n">group</span><span class="o">=</span><span class="s2">&quot;library_type&quot;</span><span class="p">,</span>
        <span class="p">),</span>
    <span class="p">)</span>
    <span class="o">+</span> <span class="n">geom_point</span><span class="p">(</span>
        <span class="n">aes</span><span class="p">(</span>
            <span class="n">fill</span><span class="o">=</span><span class="s2">&quot;mut_type&quot;</span><span class="p">,</span>
            <span class="n">shape</span><span class="o">=</span><span class="s2">&quot;library&quot;</span>
        <span class="p">),</span>
        <span class="n">size</span><span class="o">=</span><span class="mi">4</span>
    <span class="p">)</span>
    <span class="o">+</span> <span class="n">scale_fill_manual</span><span class="p">(</span><span class="n">values</span><span class="o">=</span><span class="n">CBPALETTE</span><span class="p">)</span>
    <span class="o">+</span> <span class="n">facet_wrap</span><span class="p">(</span><span class="s2">&quot;measurement_type&quot;</span><span class="p">)</span>
    <span class="o">+</span> <span class="n">theme_classic</span><span class="p">()</span>
    <span class="o">+</span> <span class="n">theme</span><span class="p">(</span>
        <span class="n">figure_size</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="mf">3.3</span><span class="p">),</span>
        <span class="n">axis_text_x</span><span class="o">=</span><span class="n">element_text</span><span class="p">(</span><span class="n">angle</span><span class="o">=</span><span class="mi">90</span><span class="p">),</span>
        <span class="n">panel_grid_major_x</span><span class="o">=</span><span class="n">element_blank</span><span class="p">(),</span>  <span class="c1"># no vertical grid lines</span>
    <span class="p">)</span>
    <span class="o">+</span> <span class="n">labs</span><span class="p">(</span>
        <span class="n">title</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;&quot;</span><span class="p">,</span>
        <span class="n">x</span><span class="o">=</span><span class="s2">&quot;lasso penalty coefficient (λ)&quot;</span><span class="p">,</span>
        <span class="n">y</span><span class="o">=</span><span class="s2">&quot;sparsity&quot;</span>
    <span class="p">)</span>
<span class="p">)</span>
<span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">(</span><span class="n">show</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/bbc14bf9f7ab6f071d777715f7e852c2533be94cf040856641a733b150b6e9b9.png" src="_images/bbc14bf9f7ab6f071d777715f7e852c2533be94cf040856641a733b150b6e9b9.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">data</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">csv_output_dir</span><span class="si">}</span><span class="s2">/fit_sparsity.csv&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">data</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>dataset_name</th>
      <th>scale_coeff_lasso_shift</th>
      <th>mut_type</th>
      <th>mut_param</th>
      <th>sparsity</th>
      <th>library</th>
      <th>library_type</th>
      <th>measurement_type</th>
      <th>True Sparsity</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>lib_1_loose_bottle_func_score</td>
      <td>0.0</td>
      <td>nonsynonymous</td>
      <td>shift_h2</td>
      <td>0.00</td>
      <td>lib_1</td>
      <td>lib_1-nonsynonymous</td>
      <td>loose_bottle</td>
      <td></td>
    </tr>
    <tr>
      <th>1</th>
      <td>lib_1_loose_bottle_func_score</td>
      <td>0.0</td>
      <td>stop</td>
      <td>shift_h2</td>
      <td>0.00</td>
      <td>lib_1</td>
      <td>lib_1-stop</td>
      <td>loose_bottle</td>
      <td></td>
    </tr>
    <tr>
      <th>2</th>
      <td>lib_1_loose_bottle_func_score</td>
      <td>0.000005</td>
      <td>nonsynonymous</td>
      <td>shift_h2</td>
      <td>0.08</td>
      <td>lib_1</td>
      <td>lib_1-nonsynonymous</td>
      <td>loose_bottle</td>
      <td></td>
    </tr>
    <tr>
      <th>3</th>
      <td>lib_1_loose_bottle_func_score</td>
      <td>0.000005</td>
      <td>stop</td>
      <td>shift_h2</td>
      <td>0.66</td>
      <td>lib_1</td>
      <td>lib_1-stop</td>
      <td>loose_bottle</td>
      <td></td>
    </tr>
    <tr>
      <th>4</th>
      <td>lib_1_loose_bottle_func_score</td>
      <td>0.00001</td>
      <td>nonsynonymous</td>
      <td>shift_h2</td>
      <td>0.15</td>
      <td>lib_1</td>
      <td>lib_1-nonsynonymous</td>
      <td>loose_bottle</td>
      <td></td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
</section>
<section id="replicate-correlations">
<h2>Replicate correlations<a class="headerlink" href="#replicate-correlations" title="Link to this heading">¶</a></h2>
<p>Another common metric to look at when evaluating models is to look at the correlation of the inferred mutational effects between the two libraries. We can do this by computing the pearson correlation coefficient between the inferred mutational effects for each mutation in the two libraries using the <code class="docutils literal notranslate"><span class="pre">multidms.ModelCollection.mut_param_dataset_correlation</span></code> <a class="reference external" href="https://matsengrp.github.io/multidms/multidms.model_collection.html#multidms.model_collection.ModelCollection.mut_param_dataset_correlation">method</a>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">chart</span><span class="p">,</span> <span class="n">data</span> <span class="o">=</span> <span class="n">model_collection</span><span class="o">.</span><span class="n">mut_param_dataset_correlation</span><span class="p">(</span><span class="n">return_data</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">data</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>datasets</th>
      <th>mut_param</th>
      <th>correlation</th>
      <th>scale_coeff_lasso_shift</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>lib_1_loose_bottle_func_score,lib_1_observed_p...</td>
      <td>beta</td>
      <td>0.990554</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>0</th>
      <td>lib_1_loose_bottle_func_score,lib_1_observed_p...</td>
      <td>beta</td>
      <td>0.992074</td>
      <td>0.000005</td>
    </tr>
    <tr>
      <th>0</th>
      <td>lib_1_loose_bottle_func_score,lib_1_observed_p...</td>
      <td>beta</td>
      <td>0.991360</td>
      <td>0.000010</td>
    </tr>
    <tr>
      <th>0</th>
      <td>lib_1_loose_bottle_func_score,lib_1_observed_p...</td>
      <td>beta</td>
      <td>0.993988</td>
      <td>0.000020</td>
    </tr>
    <tr>
      <th>0</th>
      <td>lib_1_loose_bottle_func_score,lib_1_observed_p...</td>
      <td>beta</td>
      <td>0.994166</td>
      <td>0.000040</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>By default, the method returns a dataframe giving parameter between each unique pair of datasets, meaning it includes correlations between models fit on different measurement types. We’ll only be comparing models trained the same fitting target combination.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">data</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">data</span>
    <span class="o">.</span><span class="n">assign</span><span class="p">(</span>
        <span class="n">lib1</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">datasets</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;,&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">str</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
        <span class="n">lib2</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">datasets</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;,&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">str</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span>
        <span class="n">measurement_type_1</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;lib1&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">str</span><span class="p">[</span><span class="mi">2</span><span class="p">:</span><span class="mi">4</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">),</span>
        <span class="n">measurement_type_2</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;lib2&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">str</span><span class="p">[</span><span class="mi">2</span><span class="p">:</span><span class="mi">4</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)</span>
    <span class="p">)</span>
    <span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s2">&quot;(measurement_type_1 == measurement_type_2) &amp; (~mut_param.str.contains(&#39;predicted_func_score&#39;))&quot;</span><span class="p">)</span>
    <span class="o">.</span><span class="n">rename</span><span class="p">(</span>
        <span class="p">{</span>
            <span class="s2">&quot;measurement_type_1&quot;</span> <span class="p">:</span> <span class="s2">&quot;measurement_type&quot;</span><span class="p">,</span>
        <span class="p">},</span>
        <span class="n">axis</span><span class="o">=</span><span class="mi">1</span>
    <span class="p">)</span>
    <span class="o">.</span><span class="n">replace</span><span class="p">({</span><span class="s2">&quot;shift_h2&quot;</span><span class="p">:</span> <span class="s2">&quot;shift&quot;</span><span class="p">})</span>
    <span class="o">.</span><span class="n">drop</span><span class="p">([</span><span class="s2">&quot;lib1&quot;</span><span class="p">,</span> <span class="s2">&quot;lib2&quot;</span><span class="p">,</span> <span class="s2">&quot;datasets&quot;</span><span class="p">,</span> <span class="s2">&quot;measurement_type_2&quot;</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="p">)</span>

<span class="n">data</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">10</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>mut_param</th>
      <th>correlation</th>
      <th>scale_coeff_lasso_shift</th>
      <th>measurement_type</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>beta</td>
      <td>0.979013</td>
      <td>0.000000</td>
      <td>loose_bottle</td>
    </tr>
    <tr>
      <th>0</th>
      <td>beta</td>
      <td>0.981865</td>
      <td>0.000005</td>
      <td>loose_bottle</td>
    </tr>
    <tr>
      <th>0</th>
      <td>beta</td>
      <td>0.982301</td>
      <td>0.000010</td>
      <td>loose_bottle</td>
    </tr>
    <tr>
      <th>0</th>
      <td>beta</td>
      <td>0.985200</td>
      <td>0.000020</td>
      <td>loose_bottle</td>
    </tr>
    <tr>
      <th>0</th>
      <td>beta</td>
      <td>0.986082</td>
      <td>0.000040</td>
      <td>loose_bottle</td>
    </tr>
    <tr>
      <th>0</th>
      <td>beta</td>
      <td>0.984596</td>
      <td>0.000080</td>
      <td>loose_bottle</td>
    </tr>
    <tr>
      <th>0</th>
      <td>beta</td>
      <td>0.983572</td>
      <td>0.000160</td>
      <td>loose_bottle</td>
    </tr>
    <tr>
      <th>0</th>
      <td>beta</td>
      <td>0.985226</td>
      <td>0.000320</td>
      <td>loose_bottle</td>
    </tr>
    <tr>
      <th>0</th>
      <td>beta</td>
      <td>0.985968</td>
      <td>0.000640</td>
      <td>loose_bottle</td>
    </tr>
    <tr>
      <th>0</th>
      <td>shift</td>
      <td>0.435964</td>
      <td>0.000000</td>
      <td>loose_bottle</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;measurement_type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Categorical</span><span class="p">(</span>
    <span class="n">data</span><span class="p">[</span><span class="s2">&quot;measurement_type&quot;</span><span class="p">],</span>
    <span class="n">categories</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;observed_phenotype&quot;</span><span class="p">,</span> <span class="s2">&quot;loose_bottle&quot;</span><span class="p">,</span> <span class="s2">&quot;tight_bottle&quot;</span><span class="p">],</span>
    <span class="n">ordered</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>
<span class="n">data</span><span class="p">[</span><span class="s2">&quot;mut_param&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Categorical</span><span class="p">(</span>
    <span class="n">data</span><span class="p">[</span><span class="s2">&quot;mut_param&quot;</span><span class="p">],</span>
    <span class="n">categories</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;shift&quot;</span><span class="p">,</span> <span class="s2">&quot;beta&quot;</span><span class="p">],</span>
    <span class="n">ordered</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>

<span class="n">data</span><span class="p">[</span><span class="s2">&quot;scale_coeff_lasso_shift&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">scale_coeff_lasso_shift</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">object</span><span class="p">)</span>

<span class="k">for</span> <span class="n">parameter</span><span class="p">,</span> <span class="n">parameter_df</span> <span class="ow">in</span> <span class="n">data</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;mut_param&quot;</span><span class="p">,</span> <span class="n">observed</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
    <span class="n">p</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">ggplot</span><span class="p">(</span>
            <span class="n">parameter_df</span>
        <span class="p">)</span>
        <span class="o">+</span> <span class="n">geom_line</span><span class="p">(</span>
            <span class="n">aes</span><span class="p">(</span>
                <span class="n">x</span><span class="o">=</span><span class="s2">&quot;scale_coeff_lasso_shift&quot;</span><span class="p">,</span>
                <span class="n">y</span><span class="o">=</span><span class="s2">&quot;correlation&quot;</span><span class="p">,</span>
                <span class="n">group</span><span class="o">=</span><span class="s2">&quot;measurement_type&quot;</span>
            <span class="p">),</span>
        <span class="p">)</span>
        <span class="o">+</span> <span class="n">geom_point</span><span class="p">(</span>
            <span class="n">aes</span><span class="p">(</span>
                <span class="n">x</span><span class="o">=</span><span class="s2">&quot;scale_coeff_lasso_shift&quot;</span><span class="p">,</span>
                <span class="n">y</span><span class="o">=</span><span class="s2">&quot;correlation&quot;</span><span class="p">,</span>
            <span class="p">),</span>
            <span class="n">size</span><span class="o">=</span><span class="mi">4</span>
        <span class="p">)</span>
        <span class="o">+</span> <span class="n">facet_wrap</span><span class="p">(</span><span class="s2">&quot;measurement_type&quot;</span><span class="p">,</span> <span class="n">scales</span><span class="o">=</span><span class="s2">&quot;free_y&quot;</span><span class="p">)</span>
        <span class="o">+</span> <span class="n">theme_classic</span><span class="p">()</span>
        <span class="o">+</span> <span class="n">theme</span><span class="p">(</span>
            <span class="n">figure_size</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="mf">3.3</span><span class="p">),</span>
            <span class="n">axis_text_x</span><span class="o">=</span><span class="n">element_text</span><span class="p">(</span><span class="n">angle</span><span class="o">=</span><span class="mi">90</span><span class="p">),</span>
            <span class="n">panel_grid_major_x</span><span class="o">=</span><span class="n">element_blank</span><span class="p">(),</span>  <span class="c1"># no vertical grid lines</span>
        <span class="p">)</span>
        <span class="o">+</span> <span class="n">labs</span><span class="p">(</span>
            <span class="n">title</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;Library Replicate Correlations: </span><span class="si">{</span><span class="n">parameter</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>
            <span class="n">x</span><span class="o">=</span><span class="s2">&quot;lasso penalty coefficient (λ)&quot;</span><span class="p">,</span>
            <span class="n">y</span><span class="o">=</span><span class="s2">&quot;pearsonr&quot;</span>
        <span class="p">)</span>
    <span class="p">)</span>
    <span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">(</span><span class="n">show</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/cb51f4c5a4b75783b76f5aafeeeb04895a0fcda9b0c4c1890eea7e7248f38329.png" src="_images/cb51f4c5a4b75783b76f5aafeeeb04895a0fcda9b0c4c1890eea7e7248f38329.png" />
<img alt="_images/fd5c6095ccad0a2d9a3b4f7110ca457fb627897d8843a7f9d9c337ec2193e6f8.png" src="_images/fd5c6095ccad0a2d9a3b4f7110ca457fb627897d8843a7f9d9c337ec2193e6f8.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">data</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">csv_output_dir</span><span class="si">}</span><span class="s2">/library_replicate_correlation.csv&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">data</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>mut_param</th>
      <th>correlation</th>
      <th>scale_coeff_lasso_shift</th>
      <th>measurement_type</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>beta</td>
      <td>0.98</td>
      <td>0.0</td>
      <td>loose_bottle</td>
    </tr>
    <tr>
      <th>0</th>
      <td>beta</td>
      <td>0.98</td>
      <td>0.000005</td>
      <td>loose_bottle</td>
    </tr>
    <tr>
      <th>0</th>
      <td>beta</td>
      <td>0.98</td>
      <td>0.00001</td>
      <td>loose_bottle</td>
    </tr>
    <tr>
      <th>0</th>
      <td>beta</td>
      <td>0.99</td>
      <td>0.00002</td>
      <td>loose_bottle</td>
    </tr>
    <tr>
      <th>0</th>
      <td>beta</td>
      <td>0.99</td>
      <td>0.00004</td>
      <td>loose_bottle</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
</section>
<section id="model-vs-truth-variant-phenotypes">
<h2>Model vs. truth variant phenotypes<a class="headerlink" href="#model-vs-truth-variant-phenotypes" title="Link to this heading">¶</a></h2>
<p>Let’s take a look at how the lasso constraint affects the models’ ability to predict the true latent and observed phenotypes.</p>
<p>First, we grab the model predictions of latent and observed phenotypes for each of the datasets using <code class="docutils literal notranslate"><span class="pre">multidms.Model.get_variants_df</span></code>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">variants_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span>
    <span class="p">[</span>
        <span class="n">row</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">get_variants_df</span><span class="p">(</span><span class="n">phenotype_as_effect</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="o">.</span><span class="n">assign</span><span class="p">(</span>
            <span class="n">library</span><span class="o">=</span><span class="n">row</span><span class="o">.</span><span class="n">library</span><span class="p">,</span>
            <span class="n">measurement_type</span><span class="o">=</span><span class="n">row</span><span class="o">.</span><span class="n">measurement_type</span><span class="p">,</span>
            <span class="n">scale_coeff_lasso_shift</span><span class="o">=</span><span class="n">row</span><span class="o">.</span><span class="n">scale_coeff_lasso_shift</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="o">.</span><span class="n">rename</span><span class="p">(</span>
            <span class="p">{</span>
                <span class="s2">&quot;predicted_func_score&quot;</span><span class="p">:</span> <span class="s2">&quot;predicted_phenotype&quot;</span><span class="p">,</span>
                <span class="s2">&quot;predicted_latent&quot;</span><span class="p">:</span> <span class="s2">&quot;predicted_latent_phenotype&quot;</span><span class="p">,</span>
                <span class="s2">&quot;func_score&quot;</span> <span class="p">:</span> <span class="s2">&quot;measured_phenotype&quot;</span><span class="p">,</span>
            <span class="p">},</span>
            <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="c1"># add enrichments</span>
        <span class="o">.</span><span class="n">assign</span><span class="p">(</span>
            <span class="n">predicted_enrichment</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="mi">2</span><span class="o">**</span><span class="n">x</span><span class="p">[</span><span class="s1">&#39;predicted_phenotype&#39;</span><span class="p">],</span>
            <span class="n">measured_enrichment</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="mi">2</span><span class="o">**</span><span class="n">x</span><span class="p">[</span><span class="s1">&#39;measured_phenotype&#39;</span><span class="p">],</span>
            <span class="n">fit_idx</span> <span class="o">=</span> <span class="n">idx</span>
        <span class="p">)</span>
        <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">fit_collection_df</span><span class="o">.</span><span class="n">iterrows</span><span class="p">()</span>
    <span class="p">]</span>
<span class="p">)</span>
<span class="n">variants_df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>condition</th>
      <th>aa_substitutions</th>
      <th>measured_phenotype</th>
      <th>var_wrt_ref</th>
      <th>predicted_latent_phenotype</th>
      <th>predicted_phenotype</th>
      <th>library</th>
      <th>measurement_type</th>
      <th>scale_coeff_lasso_shift</th>
      <th>predicted_enrichment</th>
      <th>measured_enrichment</th>
      <th>fit_idx</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>h1</td>
      <td>V24Q S25R H27P D29L H33C S41L C44F S47F</td>
      <td>-5.959752</td>
      <td>V24Q S25R H27P D29L H33C S41L C44F S47F</td>
      <td>-10.951781</td>
      <td>-5.962394</td>
      <td>lib_1</td>
      <td>observed_phenotype</td>
      <td>0.0</td>
      <td>0.016038</td>
      <td>0.016067</td>
      <td>0</td>
    </tr>
    <tr>
      <th>1</th>
      <td>h1</td>
      <td>V8P H27T P30* L49P</td>
      <td>-5.959753</td>
      <td>V8P H27T P30* L49P</td>
      <td>-4.826468</td>
      <td>-5.914732</td>
      <td>lib_1</td>
      <td>observed_phenotype</td>
      <td>0.0</td>
      <td>0.016576</td>
      <td>0.016067</td>
      <td>0</td>
    </tr>
    <tr>
      <th>2</th>
      <td>h1</td>
      <td>V19P V43Q</td>
      <td>-5.640393</td>
      <td>V19P V43Q</td>
      <td>-2.839960</td>
      <td>-5.630857</td>
      <td>lib_1</td>
      <td>observed_phenotype</td>
      <td>0.0</td>
      <td>0.020181</td>
      <td>0.020048</td>
      <td>0</td>
    </tr>
    <tr>
      <th>3</th>
      <td>h1</td>
      <td>R20Y</td>
      <td>0.005959</td>
      <td>R20Y</td>
      <td>5.107153</td>
      <td>0.009075</td>
      <td>lib_1</td>
      <td>observed_phenotype</td>
      <td>0.0</td>
      <td>1.006310</td>
      <td>1.004139</td>
      <td>0</td>
    </tr>
    <tr>
      <th>4</th>
      <td>h1</td>
      <td></td>
      <td>0.000000</td>
      <td></td>
      <td>4.938403</td>
      <td>0.002477</td>
      <td>lib_1</td>
      <td>observed_phenotype</td>
      <td>0.0</td>
      <td>1.001719</td>
      <td>1.000000</td>
      <td>0</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>Next, add the simulated ground truth phenotypes:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">variants_df</span>  <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span>
    <span class="p">[</span>
        <span class="n">variants_df</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s2">&quot;condition == @homolog&quot;</span><span class="p">)</span>
        <span class="o">.</span><span class="n">assign</span><span class="p">(</span>
            <span class="n">true_latent_phenotype</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;aa_substitutions&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">phenotype_fxn_dict</span><span class="p">[</span><span class="s2">&quot;latentPhenotype&quot;</span><span class="p">]),</span>
            <span class="n">true_observed_phenotype</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;aa_substitutions&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">phenotype_fxn_dict</span><span class="p">[</span><span class="s2">&quot;observedPhenotype&quot;</span><span class="p">]),</span>
            <span class="n">true_enrichment</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;aa_substitutions&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">phenotype_fxn_dict</span><span class="p">[</span><span class="s2">&quot;observedEnrichment&quot;</span><span class="p">]),</span>
        <span class="p">)</span>
        <span class="k">for</span> <span class="n">homolog</span><span class="p">,</span> <span class="n">phenotype_fxn_dict</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">([</span><span class="s2">&quot;h1&quot;</span><span class="p">,</span> <span class="s2">&quot;h2&quot;</span><span class="p">],</span> <span class="p">[</span><span class="n">phenotype_fxn_dict_h1</span><span class="p">,</span> <span class="n">phenotype_fxn_dict_h2</span><span class="p">])</span>
    <span class="p">]</span>
<span class="p">)</span>

<span class="n">variants_df</span><span class="p">[</span><span class="s2">&quot;measurement_type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Categorical</span><span class="p">(</span>
    <span class="n">variants_df</span><span class="p">[</span><span class="s2">&quot;measurement_type&quot;</span><span class="p">],</span>
    <span class="n">categories</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;observed_phenotype&quot;</span><span class="p">,</span> <span class="s2">&quot;loose_bottle&quot;</span><span class="p">,</span> <span class="s2">&quot;tight_bottle&quot;</span><span class="p">],</span>
    <span class="n">ordered</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>

<span class="c1"># As a sanity check let&#39;s make sure the true phenotypes match the </span>
<span class="c1"># &quot;measured phenotypes&quot; (i.e. fitting targets) for the variants that were trained on ground truth targets</span>
<span class="k">assert</span> <span class="n">np</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span>
    <span class="n">variants_df</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s2">&quot;measurement_type == &#39;observed_phenotype&#39;&quot;</span><span class="p">)[</span><span class="s1">&#39;measured_phenotype&#39;</span><span class="p">],</span>
    <span class="n">variants_df</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s2">&quot;measurement_type == &#39;observed_phenotype&#39;&quot;</span><span class="p">)[</span><span class="s1">&#39;true_observed_phenotype&#39;</span><span class="p">],</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>For clarity, let’s define and arrange the columns we now have:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cols</span> <span class="o">=</span> <span class="p">[</span>
    <span class="c1"># unique combinations of these make up the model collection that we&#39;ve fit</span>
    <span class="s2">&quot;library&quot;</span><span class="p">,</span> <span class="c1"># replicate library,</span>
    <span class="s2">&quot;measurement_type&quot;</span><span class="p">,</span> <span class="c1"># type of functional score</span>
    <span class="s2">&quot;scale_coeff_lasso_shift&quot;</span><span class="p">,</span> <span class="c1"># lasso coefficient of model</span>

    <span class="c1"># variant defining columns</span>
    <span class="s2">&quot;aa_substitutions&quot;</span><span class="p">,</span> <span class="c1"># variant substitutions</span>
    <span class="s2">&quot;var_wrt_ref&quot;</span><span class="p">,</span> <span class="c1"># variant substitutions relative to reference wildtype</span>
    <span class="s2">&quot;condition&quot;</span><span class="p">,</span> <span class="c1"># homolog</span>

    <span class="s2">&quot;measured_phenotype&quot;</span><span class="p">,</span> <span class="c1"># the actual target functional score for training, in multidms, this is &quot;func_score&quot;</span>
    <span class="s2">&quot;measured_enrichment&quot;</span><span class="p">,</span> <span class="c1"># 2 ** measured_func_score</span>

    <span class="s2">&quot;predicted_latent_phenotype&quot;</span><span class="p">,</span> <span class="c1"># predicted latent phenotype</span>
    <span class="s2">&quot;predicted_phenotype&quot;</span><span class="p">,</span> <span class="c1"># predicted observed phenotype - or in jesse&#39;s case, the &quot;observed phenotype&quot;</span>
    <span class="s2">&quot;predicted_enrichment&quot;</span><span class="p">,</span> <span class="c1"># predicted enrichment</span>

    <span class="s2">&quot;true_latent_phenotype&quot;</span><span class="p">,</span> <span class="c1"># true latent phenotype</span>
    <span class="s2">&quot;true_observed_phenotype&quot;</span><span class="p">,</span> <span class="c1"># true observed phenotype</span>
    <span class="s2">&quot;true_enrichment&quot;</span><span class="p">,</span> <span class="c1"># true enrichment</span>
<span class="p">]</span>
<span class="c1"># variants_df[cols].round(2).head()</span>
</pre></div>
</div>
</div>
</div>
<p>Add correlation of ground truth targets / predicted phenotypes to the fit collection dataframe and plot them:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">model_variants_df</span> <span class="ow">in</span> <span class="n">variants_df</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;fit_idx&quot;</span><span class="p">):</span>
    <span class="k">for</span> <span class="n">metric_fxn</span><span class="p">,</span> <span class="n">metric_name</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">([</span><span class="n">series_corr</span><span class="p">,</span> <span class="n">series_mae</span><span class="p">],</span> <span class="p">[</span><span class="s2">&quot;corr&quot;</span><span class="p">,</span> <span class="s2">&quot;mae&quot;</span><span class="p">]):</span>
        <span class="n">fit_collection_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">idx</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;variant_phenotype_</span><span class="si">{</span><span class="n">metric_name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">metric_fxn</span><span class="p">(</span>
            <span class="n">model_variants_df</span><span class="p">[</span><span class="s2">&quot;measured_phenotype&quot;</span><span class="p">],</span>
            <span class="n">model_variants_df</span><span class="p">[</span><span class="s2">&quot;predicted_phenotype&quot;</span><span class="p">]</span>
        <span class="p">)</span>

<span class="n">fit_collection_df</span><span class="p">[[</span><span class="s2">&quot;scale_coeff_lasso_shift&quot;</span><span class="p">,</span> <span class="s2">&quot;dataset_name&quot;</span><span class="p">,</span> <span class="s2">&quot;variant_phenotype_corr&quot;</span><span class="p">,</span> <span class="s2">&quot;variant_phenotype_mae&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>scale_coeff_lasso_shift</th>
      <th>dataset_name</th>
      <th>variant_phenotype_corr</th>
      <th>variant_phenotype_mae</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0.0</td>
      <td>lib_1_observed_phenotype_func_score</td>
      <td>0.999985</td>
      <td>8.611429e-06</td>
    </tr>
    <tr>
      <th>1</th>
      <td>0.000005</td>
      <td>lib_1_observed_phenotype_func_score</td>
      <td>0.999994</td>
      <td>5.973305e-07</td>
    </tr>
    <tr>
      <th>2</th>
      <td>0.00001</td>
      <td>lib_1_observed_phenotype_func_score</td>
      <td>0.999989</td>
      <td>1.648540e-04</td>
    </tr>
    <tr>
      <th>3</th>
      <td>0.00002</td>
      <td>lib_1_observed_phenotype_func_score</td>
      <td>0.999980</td>
      <td>1.652058e-04</td>
    </tr>
    <tr>
      <th>4</th>
      <td>0.00004</td>
      <td>lib_1_observed_phenotype_func_score</td>
      <td>0.999950</td>
      <td>3.538619e-06</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># for metric in [&quot;corr&quot;, &quot;mae&quot;]:</span>
<span class="n">data</span> <span class="o">=</span> <span class="n">fit_collection_df</span><span class="p">[[</span><span class="s2">&quot;library&quot;</span><span class="p">,</span> <span class="s2">&quot;measurement_type&quot;</span><span class="p">,</span> <span class="s2">&quot;scale_coeff_lasso_shift&quot;</span><span class="p">,</span> <span class="s2">&quot;dataset_name&quot;</span><span class="p">,</span> <span class="s2">&quot;variant_phenotype_corr&quot;</span><span class="p">,</span> <span class="s2">&quot;variant_phenotype_mae&quot;</span><span class="p">]]</span>
<span class="n">metric</span> <span class="o">=</span> <span class="s2">&quot;corr&quot;</span>
<span class="n">p</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">ggplot</span><span class="p">(</span>
        <span class="n">data</span>
        <span class="o">.</span><span class="n">assign</span><span class="p">(</span>
            <span class="n">measurement_library</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;library&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;measurement_type&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
        <span class="p">)</span>
    <span class="p">)</span>
    <span class="o">+</span> <span class="n">geom_line</span><span class="p">(</span>
        <span class="n">aes</span><span class="p">(</span>
            <span class="n">x</span><span class="o">=</span><span class="s2">&quot;scale_coeff_lasso_shift&quot;</span><span class="p">,</span>
            <span class="n">y</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;variant_phenotype_</span><span class="si">{</span><span class="n">metric</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>
            <span class="n">group</span><span class="o">=</span><span class="s2">&quot;measurement_library&quot;</span><span class="p">,</span>
        <span class="p">),</span>
    <span class="p">)</span>
    <span class="o">+</span> <span class="n">geom_point</span><span class="p">(</span>
        <span class="n">aes</span><span class="p">(</span>
            <span class="n">x</span><span class="o">=</span><span class="s2">&quot;scale_coeff_lasso_shift&quot;</span><span class="p">,</span>
            <span class="n">y</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;variant_phenotype_</span><span class="si">{</span><span class="n">metric</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>
            <span class="n">shape</span><span class="o">=</span><span class="s2">&quot;library&quot;</span>
        <span class="p">),</span>
        <span class="n">size</span><span class="o">=</span><span class="mi">4</span>
    <span class="p">)</span>
    <span class="o">+</span> <span class="n">facet_wrap</span><span class="p">(</span><span class="s2">&quot;measurement_type&quot;</span><span class="p">,</span> <span class="n">scales</span><span class="o">=</span><span class="s2">&quot;free_y&quot;</span><span class="p">)</span>
    <span class="o">+</span> <span class="n">theme_classic</span><span class="p">()</span>
    <span class="o">+</span> <span class="n">theme</span><span class="p">(</span>
        <span class="n">figure_size</span><span class="o">=</span><span class="p">(</span><span class="mf">6.5</span><span class="p">,</span> <span class="mf">3.3</span><span class="p">),</span>
        <span class="n">axis_text_x</span><span class="o">=</span><span class="n">element_text</span><span class="p">(</span><span class="n">angle</span><span class="o">=</span><span class="mi">90</span><span class="p">),</span>
        <span class="n">panel_grid_major_x</span><span class="o">=</span><span class="n">element_blank</span><span class="p">(),</span>  <span class="c1"># no vertical grid lines</span>
    <span class="p">)</span>
    <span class="o">+</span> <span class="n">labs</span><span class="p">(</span>
        <span class="n">title</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;Predicted Vs. True Variant Phenotype&quot;</span><span class="p">,</span>
        <span class="n">x</span><span class="o">=</span><span class="s2">&quot;lasso penalty coefficient ($\lambda_</span><span class="si">{L1}</span><span class="s2">$)&quot;</span><span class="p">,</span>
        <span class="n">y</span><span class="o">=</span><span class="s2">&quot;pearsonr&quot;</span> <span class="k">if</span> <span class="n">metric</span> <span class="o">==</span> <span class="s2">&quot;corr&quot;</span> <span class="k">else</span> <span class="s2">&quot;mean absolute error&quot;</span>
    <span class="p">)</span>
<span class="p">)</span>
<span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">(</span><span class="n">show</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/6f85da9e9efe826123e796667b90c911e8b01627083eba9d37a915461ab80676.png" src="_images/6f85da9e9efe826123e796667b90c911e8b01627083eba9d37a915461ab80676.png" />
</div>
</div>
<p>As expected, the lasso constraint has a negative effect on the models’ ability to predict the true latent and observed phenotypes. This is because a model with more freedom to choose parameters is likely to overfit to the training data.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">data</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">csv_output_dir</span><span class="si">}</span><span class="s2">/model_vs_truth_variant_phenotype.csv&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">data</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>library</th>
      <th>measurement_type</th>
      <th>scale_coeff_lasso_shift</th>
      <th>dataset_name</th>
      <th>variant_phenotype_corr</th>
      <th>variant_phenotype_mae</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>lib_1</td>
      <td>observed_phenotype</td>
      <td>0.0</td>
      <td>lib_1_observed_phenotype_func_score</td>
      <td>1.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>1</th>
      <td>lib_1</td>
      <td>observed_phenotype</td>
      <td>0.000005</td>
      <td>lib_1_observed_phenotype_func_score</td>
      <td>1.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>2</th>
      <td>lib_1</td>
      <td>observed_phenotype</td>
      <td>0.00001</td>
      <td>lib_1_observed_phenotype_func_score</td>
      <td>1.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>3</th>
      <td>lib_1</td>
      <td>observed_phenotype</td>
      <td>0.00002</td>
      <td>lib_1_observed_phenotype_func_score</td>
      <td>1.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>4</th>
      <td>lib_1</td>
      <td>observed_phenotype</td>
      <td>0.00004</td>
      <td>lib_1_observed_phenotype_func_score</td>
      <td>1.0</td>
      <td>0.0</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
</section>
<section id="cross-validation">
<h2>Cross validation<a class="headerlink" href="#cross-validation" title="Link to this heading">¶</a></h2>
<p>Above, we saw evidence that the lasso negatively impacts the model performance on the training data. To test for overfitting, we can perform cross validation to test that the model is actually more accurate on unseen data.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">train</span><span class="p">,</span> <span class="n">test</span> <span class="o">=</span> <span class="p">[],</span> <span class="p">{}</span>
<span class="k">for</span> <span class="p">(</span><span class="n">library</span><span class="p">,</span> <span class="n">measurement</span><span class="p">),</span> <span class="n">fs_df</span> <span class="ow">in</span> <span class="n">func_scores</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;homolog&quot;</span><span class="p">:</span><span class="s2">&quot;condition&quot;</span><span class="p">})</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s2">&quot;library&quot;</span><span class="p">,</span> <span class="s2">&quot;func_score_type&quot;</span><span class="p">]):</span>
    
    <span class="k">if</span> <span class="s2">&quot;enrichment&quot;</span> <span class="ow">in</span> <span class="n">measurement</span><span class="p">:</span>
        <span class="k">continue</span>

    <span class="n">dataset</span> <span class="o">=</span> <span class="n">fs_df</span><span class="o">.</span><span class="n">sample</span><span class="p">(</span><span class="n">frac</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="n">seed</span><span class="p">)</span>
    <span class="n">train_split</span><span class="p">,</span> <span class="n">test_split</span> <span class="o">=</span> <span class="n">dataset</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:</span><span class="nb">int</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">dataset</span><span class="p">)</span> <span class="o">*</span> <span class="n">train_frac</span><span class="p">)],</span> <span class="n">dataset</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">dataset</span><span class="p">)</span> <span class="o">*</span> <span class="n">train_frac</span><span class="p">):]</span>
    <span class="n">name</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">library</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">measurement</span><span class="si">}</span><span class="s2">&quot;</span>
    
    <span class="n">train</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
        <span class="n">multidms</span><span class="o">.</span><span class="n">Data</span><span class="p">(</span>
            <span class="n">train_split</span><span class="p">,</span> 
            <span class="n">reference</span><span class="o">=</span><span class="s2">&quot;h1&quot;</span><span class="p">,</span> 
            <span class="n">alphabet</span> <span class="o">=</span> <span class="n">multidms</span><span class="o">.</span><span class="n">AAS_WITHSTOP_WITHGAP</span><span class="p">,</span> 
            <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
            <span class="n">name</span> <span class="o">=</span> <span class="n">name</span>
        <span class="p">)</span>
    <span class="p">)</span>

    <span class="n">test</span><span class="p">[</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="n">test_split</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fitting_params</span><span class="p">[</span><span class="s2">&quot;dataset&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">train</span> 
<span class="n">_</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">fit_collection_cv</span> <span class="o">=</span> <span class="n">multidms</span><span class="o">.</span><span class="n">model_collection</span><span class="o">.</span><span class="n">fit_models</span><span class="p">(</span><span class="n">fitting_params</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">multidms.ModelCollection.add_validation_loss</span></code> method takes in unseen data, computes model loss on that data, and appends it to the <code class="docutils literal notranslate"><span class="pre">ModelCollection.fit_models</span></code> dataframe.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mc</span> <span class="o">=</span> <span class="n">multidms</span><span class="o">.</span><span class="n">model_collection</span><span class="o">.</span><span class="n">ModelCollection</span><span class="p">(</span><span class="n">fit_collection_cv</span><span class="p">)</span>
<span class="n">mc</span><span class="o">.</span><span class="n">add_validation_loss</span><span class="p">(</span><span class="n">test</span><span class="p">,</span> <span class="n">overwrite</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mc</span><span class="o">.</span><span class="n">fit_models</span><span class="p">[[</span><span class="n">c</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">mc</span><span class="o">.</span><span class="n">fit_models</span><span class="o">.</span><span class="n">columns</span> <span class="k">if</span> <span class="s2">&quot;loss&quot;</span> <span class="ow">in</span> <span class="n">c</span><span class="p">]]</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>step_loss</th>
      <th>h1_loss_training</th>
      <th>h2_loss_training</th>
      <th>total_loss_training</th>
      <th>h1_loss_validation</th>
      <th>h2_loss_validation</th>
      <th>total_loss_validation</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>[6.3576507568359375, 0.00012060716835549101]</td>
      <td>0.000078</td>
      <td>0.000043</td>
      <td>0.000121</td>
      <td>0.001116</td>
      <td>0.000479</td>
      <td>0.001595</td>
    </tr>
    <tr>
      <th>1</th>
      <td>[6.3576507568359375, 8.627029455965385e-05]</td>
      <td>0.000032</td>
      <td>0.000054</td>
      <td>0.000086</td>
      <td>0.000089</td>
      <td>0.000804</td>
      <td>0.000893</td>
    </tr>
    <tr>
      <th>2</th>
      <td>[6.3576507568359375, 0.00013628264423459768]</td>
      <td>0.000053</td>
      <td>0.000083</td>
      <td>0.000136</td>
      <td>0.000207</td>
      <td>0.001059</td>
      <td>0.001266</td>
    </tr>
    <tr>
      <th>3</th>
      <td>[6.3576507568359375, 0.000258071580901742]</td>
      <td>0.000117</td>
      <td>0.000141</td>
      <td>0.000258</td>
      <td>0.000618</td>
      <td>0.001297</td>
      <td>0.001914</td>
    </tr>
    <tr>
      <th>4</th>
      <td>[6.3576507568359375, 0.0006348328315652907]</td>
      <td>0.000305</td>
      <td>0.000330</td>
      <td>0.000635</td>
      <td>0.001270</td>
      <td>0.001042</td>
      <td>0.002312</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">data</span> <span class="o">=</span> <span class="n">mc</span><span class="o">.</span><span class="n">fit_models</span>

<span class="n">data</span><span class="p">[</span><span class="s2">&quot;library&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;dataset_name&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">str</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)</span>
<span class="n">data</span><span class="p">[</span><span class="s2">&quot;measurement_type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;dataset_name&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">str</span><span class="p">[</span><span class="mi">2</span><span class="p">:</span><span class="mi">4</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)</span>
<span class="n">data</span><span class="p">[</span><span class="s2">&quot;measurement_type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Categorical</span><span class="p">(</span>
    <span class="n">data</span><span class="p">[</span><span class="s2">&quot;measurement_type&quot;</span><span class="p">],</span>
    <span class="n">categories</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;observed_phenotype&quot;</span><span class="p">,</span> <span class="s2">&quot;loose_bottle&quot;</span><span class="p">,</span> <span class="s2">&quot;tight_bottle&quot;</span><span class="p">],</span>
    <span class="n">ordered</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>

<span class="n">data</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">data</span>
    <span class="o">.</span><span class="n">melt</span><span class="p">(</span>
        <span class="n">id_vars</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;scale_coeff_lasso_shift&quot;</span><span class="p">,</span> <span class="s2">&quot;library&quot;</span><span class="p">,</span> <span class="s2">&quot;measurement_type&quot;</span><span class="p">],</span>
        <span class="n">value_vars</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;total_loss_training&quot;</span><span class="p">,</span> <span class="s2">&quot;total_loss_validation&quot;</span><span class="p">],</span>
        <span class="n">var_name</span><span class="o">=</span><span class="s2">&quot;dataset&quot;</span><span class="p">,</span>
        <span class="n">value_name</span><span class="o">=</span><span class="s2">&quot;loss&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="o">.</span><span class="n">assign</span><span class="p">(</span>
        <span class="n">dataset</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;dataset&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;_loss&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
    <span class="p">)</span>
    <span class="o">.</span><span class="n">assign</span><span class="p">(</span>
        <span class="n">lib_dataset</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;library&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="s2">&quot; &quot;</span> <span class="o">+</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;dataset&quot;</span><span class="p">]</span>
    <span class="p">)</span>
<span class="p">)</span>
<span class="n">p</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">ggplot</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
    <span class="o">+</span> <span class="n">geom_line</span><span class="p">(</span>
        <span class="n">aes</span><span class="p">(</span>
            <span class="n">x</span><span class="o">=</span><span class="s2">&quot;scale_coeff_lasso_shift&quot;</span><span class="p">,</span>
            <span class="n">y</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;loss&quot;</span><span class="p">,</span>
            <span class="n">group</span><span class="o">=</span><span class="s2">&quot;lib_dataset&quot;</span><span class="p">,</span>
        <span class="p">)</span>
    <span class="p">)</span>
    <span class="o">+</span> <span class="n">geom_point</span><span class="p">(</span>
        <span class="n">aes</span><span class="p">(</span>
            <span class="n">x</span><span class="o">=</span><span class="s2">&quot;scale_coeff_lasso_shift&quot;</span><span class="p">,</span>
            <span class="n">y</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;loss&quot;</span><span class="p">,</span>
            <span class="n">fill</span><span class="o">=</span><span class="s2">&quot;dataset&quot;</span><span class="p">,</span>
            <span class="n">shape</span><span class="o">=</span><span class="s2">&quot;library&quot;</span>
        <span class="p">),</span>
        <span class="n">size</span><span class="o">=</span><span class="mi">4</span>
    <span class="p">)</span>
    <span class="o">+</span> <span class="n">facet_wrap</span><span class="p">(</span><span class="s2">&quot;measurement_type&quot;</span><span class="p">,</span> <span class="n">scales</span><span class="o">=</span><span class="s2">&quot;free_y&quot;</span><span class="p">)</span>
    <span class="o">+</span> <span class="n">theme_classic</span><span class="p">()</span>
    <span class="o">+</span> <span class="n">theme</span><span class="p">(</span>
        <span class="n">figure_size</span><span class="o">=</span><span class="p">(</span><span class="mf">6.5</span><span class="p">,</span> <span class="mf">3.3</span><span class="p">),</span>
        <span class="n">axis_text_x</span><span class="o">=</span><span class="n">element_text</span><span class="p">(</span><span class="n">angle</span><span class="o">=</span><span class="mi">90</span><span class="p">),</span>
        <span class="n">panel_grid_major_x</span><span class="o">=</span><span class="n">element_blank</span><span class="p">(),</span>  <span class="c1"># no vertical grid lines</span>
    <span class="p">)</span>
    <span class="o">+</span> <span class="n">labs</span><span class="p">(</span>
        <span class="n">x</span><span class="o">=</span><span class="s2">&quot;lasso penalty coefficient (λ)&quot;</span><span class="p">,</span>
        <span class="n">y</span><span class="o">=</span><span class="s2">&quot;Huber loss w/o penalty&quot;</span>
    <span class="p">)</span>
<span class="p">)</span>
<span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">(</span><span class="n">show</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/2116323723318eb7a1939df9e1e79a1b7826d936d0e803325041287e1a5a6774.png" src="_images/2116323723318eb7a1939df9e1e79a1b7826d936d0e803325041287e1a5a6774.png" />
</div>
</div>
<p>Above we can see that indeed for both the loose and tight bottleneck datasets, the lasso constraint provides a more robust model to unseen data. We don’t however see this effect for the models trained on ground truth observed phenotypes, as there is no noise for the models to overfit to.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">data</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">csv_output_dir</span><span class="si">}</span><span class="s2">/cross_validation_loss.csv&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">data</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>scale_coeff_lasso_shift</th>
      <th>library</th>
      <th>measurement_type</th>
      <th>dataset</th>
      <th>loss</th>
      <th>lib_dataset</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0.0</td>
      <td>lib_1</td>
      <td>observed_phenotype</td>
      <td>total_training</td>
      <td>0.0</td>
      <td>lib_1 total_training</td>
    </tr>
    <tr>
      <th>1</th>
      <td>0.000005</td>
      <td>lib_1</td>
      <td>observed_phenotype</td>
      <td>total_training</td>
      <td>0.0</td>
      <td>lib_1 total_training</td>
    </tr>
    <tr>
      <th>2</th>
      <td>0.00001</td>
      <td>lib_1</td>
      <td>observed_phenotype</td>
      <td>total_training</td>
      <td>0.0</td>
      <td>lib_1 total_training</td>
    </tr>
    <tr>
      <th>3</th>
      <td>0.00002</td>
      <td>lib_1</td>
      <td>observed_phenotype</td>
      <td>total_training</td>
      <td>0.0</td>
      <td>lib_1 total_training</td>
    </tr>
    <tr>
      <th>4</th>
      <td>0.00004</td>
      <td>lib_1</td>
      <td>observed_phenotype</td>
      <td>total_training</td>
      <td>0.0</td>
      <td>lib_1 total_training</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
</section>
<section id="final-selection">
<h2>Final Selection<a class="headerlink" href="#final-selection" title="Link to this heading">¶</a></h2>
<p>This analysis above provides a number of clues to inform our choice of model. With empricial data we obviously won’t have comparisons to the ground truth values, and thus the choice of a lasso penalty will laregely depend on the shift parameter sparsity, correlation of inferred mutational effects between replicate libraries, and cross validation performance. Focusing on the <em>loose bottleneck</em> training dataset (a fairly realistic level of noise that we often observe in real experiments), it seems that a lasso penalty of $1e-4$ provides a false positive rate of nearly zero (i.e. the stop codon sparsity $\approx 100%$), a good correlation of inferred mutational effects between replicate libraries (&gt;0.95 pearsonr), and a relatively low loss on validation when compared to the other models.</p>
<p>To validate the ability to correctly infer the shape of epistasis, we’ll plot the inferred global epistasis function for each of the models, and compare it to the true global epistasis function.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">data</span> <span class="o">=</span> <span class="n">variants_df</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;scale_coeff_lasso_shift == </span><span class="si">{</span><span class="n">lasso_choice</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span> <span class="ow">in</span> <span class="n">itertools</span><span class="o">.</span><span class="n">combinations</span><span class="p">(</span>
    <span class="p">[</span><span class="s2">&quot;predicted_latent_phenotype&quot;</span><span class="p">,</span> <span class="s2">&quot;predicted_phenotype&quot;</span><span class="p">,</span> <span class="s2">&quot;measured_phenotype&quot;</span><span class="p">],</span> <span class="mi">2</span>
<span class="p">):</span>
    <span class="n">p</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">ggplot</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">sample</span><span class="p">(</span><span class="n">frac</span><span class="o">=</span><span class="mf">1.0</span><span class="p">))</span>
        <span class="o">+</span> <span class="n">geom_point</span><span class="p">(</span>
            <span class="n">aes</span><span class="p">(</span>
                <span class="n">x</span><span class="o">=</span><span class="n">x</span><span class="p">,</span>
                <span class="n">y</span><span class="o">=</span><span class="n">y</span>    
            <span class="p">),</span>
            <span class="n">alpha</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span>
            <span class="n">color</span><span class="o">=</span><span class="s2">&quot;grey&quot;</span>
        <span class="p">)</span>
        <span class="o">+</span> <span class="n">ylab</span><span class="p">(</span><span class="n">y</span><span class="p">)</span>
        <span class="o">+</span> <span class="n">xlab</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>
        <span class="o">+</span> <span class="n">facet_grid</span><span class="p">(</span><span class="s2">&quot;~measurement_type&quot;</span><span class="p">)</span>
        <span class="o">+</span> <span class="n">theme_classic</span><span class="p">()</span>
        <span class="o">+</span> <span class="n">theme</span><span class="p">(</span>
            <span class="n">figure_size</span><span class="o">=</span><span class="p">(</span><span class="mf">6.5</span><span class="p">,</span> <span class="mf">3.5</span><span class="p">),</span>
            <span class="n">axis_text_x</span><span class="o">=</span><span class="n">element_text</span><span class="p">(</span><span class="n">angle</span><span class="o">=</span><span class="mi">90</span><span class="p">),</span>
        <span class="p">)</span>
    <span class="p">)</span>

    <span class="k">if</span> <span class="n">x</span> <span class="o">==</span> <span class="s2">&quot;predicted_latent_phenotype&quot;</span><span class="p">:</span>
        <span class="n">p</span><span class="o">+=</span> <span class="n">geom_point</span><span class="p">(</span>
            <span class="n">aes</span><span class="p">(</span>
                <span class="n">x</span><span class="o">=</span><span class="s2">&quot;true_latent_phenotype&quot;</span><span class="p">,</span>
                <span class="n">y</span><span class="o">=</span><span class="s2">&quot;true_observed_phenotype&quot;</span>
            <span class="p">),</span>
            <span class="n">size</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span>
        <span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">p</span> <span class="o">+=</span> <span class="n">geom_abline</span><span class="p">(</span>
            <span class="n">slope</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
            <span class="n">intercept</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> 
        <span class="p">)</span>
        
    <span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">(</span><span class="n">show</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/7d77bcf3fd3c64baad34cdd484cf775d26e7d0b9e8867c9508927dad618cf1cb.png" src="_images/7d77bcf3fd3c64baad34cdd484cf775d26e7d0b9e8867c9508927dad618cf1cb.png" />
<img alt="_images/51c0da65cdc1a73dab7114ab3777ebddfd620fad1b4714f17c7686f220445a44.png" src="_images/51c0da65cdc1a73dab7114ab3777ebddfd620fad1b4714f17c7686f220445a44.png" />
<img alt="_images/d3e42c71ba009b7574f1aeef8b0ff72c178bf66e983b2ac6982d862b9bf6fdf0.png" src="_images/d3e42c71ba009b7574f1aeef8b0ff72c178bf66e983b2ac6982d862b9bf6fdf0.png" />
</div>
</div>
<p>Plot model vs. measured (functional score) vs. ground truth enrichments.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">p</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">ggplot</span><span class="p">(</span>
        <span class="n">data</span><span class="o">.</span><span class="n">melt</span><span class="p">(</span>
            <span class="c1"># id_vars=[&quot;measurement_type&quot;, &quot;library&quot;, &quot;condition&quot;, &quot;true_enrichment&quot;],</span>
            <span class="n">id_vars</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;measurement_type&quot;</span><span class="p">,</span> <span class="s2">&quot;library&quot;</span><span class="p">,</span> <span class="s2">&quot;true_enrichment&quot;</span><span class="p">],</span>
            <span class="n">value_vars</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;predicted_enrichment&quot;</span><span class="p">,</span> <span class="s2">&quot;measured_enrichment&quot;</span><span class="p">],</span>
            <span class="n">var_name</span><span class="o">=</span><span class="s2">&quot;enrichment_type&quot;</span><span class="p">,</span>
            <span class="n">value_name</span><span class="o">=</span><span class="s2">&quot;enrichment&quot;</span><span class="p">,</span>
        <span class="p">),</span>
        <span class="c1"># variants_df,</span>
        <span class="n">aes</span><span class="p">(</span><span class="s2">&quot;true_enrichment&quot;</span><span class="p">,</span> <span class="s2">&quot;enrichment&quot;</span><span class="p">),</span>
    <span class="p">)</span>
    <span class="o">+</span> <span class="n">geom_abline</span><span class="p">(</span>
        <span class="n">slope</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
        <span class="n">intercept</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> 
        <span class="c1"># alpha=0.5,</span>
        <span class="c1"># size=0.5,</span>
    <span class="p">)</span>
    <span class="o">+</span> <span class="n">geom_point</span><span class="p">(</span><span class="n">alpha</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
    <span class="o">+</span> <span class="n">facet_grid</span><span class="p">(</span><span class="s2">&quot;enrichment_type~measurement_type&quot;</span><span class="p">,</span> <span class="n">scales</span><span class="o">=</span><span class="s2">&quot;free_y&quot;</span><span class="p">)</span>
    <span class="o">+</span> <span class="n">theme_classic</span><span class="p">()</span>
    <span class="o">+</span> <span class="n">theme</span><span class="p">(</span>
        <span class="n">figure_size</span><span class="o">=</span><span class="p">(</span><span class="mf">6.5</span><span class="p">,</span> <span class="mi">5</span><span class="p">),</span>
    <span class="p">)</span>
<span class="p">)</span>

<span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">(</span><span class="n">show</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/278028a4883fa86828826275c78807089cbb61a18b07e7a101b3dbf3fb2d89f3.png" src="_images/278028a4883fa86828826275c78807089cbb61a18b07e7a101b3dbf3fb2d89f3.png" />
</div>
</div>
<p>Amazingly, the model does better at predicting true enrichments than even the counts based functional scores!</p>
<p>Finally, let’s plot the relationship between the model’s inferred shifts, and the ground truth shifts.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># plot correlations of true and predicted shifts</span>
<span class="n">collection_muts_df</span><span class="p">[</span><span class="s2">&quot;measurement_type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Categorical</span><span class="p">(</span>    
    <span class="n">collection_muts_df</span><span class="p">[</span><span class="s2">&quot;measurement_type&quot;</span><span class="p">],</span>
    <span class="n">categories</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;observed_phenotype&quot;</span><span class="p">,</span> <span class="s2">&quot;loose_bottle&quot;</span><span class="p">,</span> <span class="s2">&quot;tight_bottle&quot;</span><span class="p">],</span>
    <span class="n">ordered</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>

<span class="k">for</span> <span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">)</span> <span class="ow">in</span> <span class="n">itertools</span><span class="o">.</span><span class="n">combinations</span><span class="p">(</span>
    <span class="p">[</span><span class="s2">&quot;true_shift&quot;</span><span class="p">,</span> <span class="s2">&quot;predicted_shift_h2&quot;</span><span class="p">],</span> <span class="mi">2</span>
<span class="p">):</span>
    <span class="n">p</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">ggplot</span><span class="p">(</span><span class="n">collection_muts_df</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;scale_coeff_lasso_shift == </span><span class="si">{</span><span class="n">lasso_choice</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">))</span>
        <span class="o">+</span> <span class="n">geom_point</span><span class="p">(</span>
            <span class="n">aes</span><span class="p">(</span>
                <span class="n">x</span><span class="o">=</span><span class="n">x</span><span class="p">,</span>
                <span class="n">y</span><span class="o">=</span><span class="n">y</span>    
            <span class="p">),</span>
            <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="o">+</span> <span class="n">geom_abline</span><span class="p">(</span>
            <span class="n">slope</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
            <span class="n">intercept</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> 
        <span class="p">)</span>
        <span class="o">+</span> <span class="n">ylab</span><span class="p">(</span><span class="n">y</span><span class="p">)</span>
        <span class="o">+</span> <span class="n">xlab</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>
        <span class="o">+</span> <span class="n">facet_grid</span><span class="p">(</span><span class="s2">&quot;library~measurement_type&quot;</span><span class="p">)</span>
        <span class="o">+</span> <span class="n">theme_classic</span><span class="p">()</span>
        <span class="o">+</span> <span class="n">theme</span><span class="p">(</span>
            <span class="n">figure_size</span><span class="o">=</span><span class="p">(</span><span class="mf">6.5</span><span class="p">,</span> <span class="mi">5</span><span class="p">),</span>
            <span class="n">axis_text_x</span><span class="o">=</span><span class="n">element_text</span><span class="p">(</span><span class="n">angle</span><span class="o">=</span><span class="mi">90</span><span class="p">),</span>
        <span class="p">)</span>
    <span class="p">)</span>
    <span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">(</span><span class="n">show</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/8ce9d3298908c07176c51c0e2bc6403d4ab31d4f5210e538e27f5cc95d9ffb4b.png" src="_images/8ce9d3298908c07176c51c0e2bc6403d4ab31d4f5210e538e27f5cc95d9ffb4b.png" />
</div>
</div>
</section>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">Joint Modeling of SARS-CoV-2 Spike homologs</a></h1>








<h3>Navigation</h3>
<ul>
<li class="toctree-l1"><a class="reference internal" href="index.html">Overview</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="spike-analysis.html">Spike Analysis</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>document.getElementById('searchbox').style.display = "block"</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &#169;2023, Jared Galloway.
      
      |
      Powered by <a href="https://www.sphinx-doc.org/">Sphinx 7.2.6</a>
      &amp; <a href="https://alabaster.readthedocs.io">Alabaster 0.7.16</a>
      
      |
      <a href="_sources/simulation_validation.ipynb.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>